// @financial-times/vvc/src/components/axis/styles.js
/**
 * @file
 * CSS styles for <Axis />
 */

import { css } from '@emotion/core';

export const baseAxisStyles = css`
  font-size: inherit !important;
`;

export const webAxisStyles = css`
  ${baseAxisStyles}

  text {
    font-family: metric, MetricWeb;
    fill: #66605c;
    font-size: 16px;
  }

  line {
    stroke: #e6d9ce;
    stroke-width: 1px;
  }

  .visx-axis-banding rect {
    fill: #ccc1b7;
    opacity: 0.4;
  }

  /* tick text on yAxis */
  .visx-axis-left text,
  .visx-axis-right text {
    fill: #66605c;
  }

  .visx-axis-left text {
    text-anchor: end;
  }

  .visx-axis-right text {
    text-anchor: start;
  }

  /* tick line on yAxis */
  .visx-axis-left line,
  .visx-axis-right line {
    stroke: #e6d9ce;
    stroke-width: 1px;
  }

  /* tick text on xAxis */
  .visx-axis-top text,
  .visx-axis-bottom text {
    fill: #66605c;
    text-anchor: middle;
  }

  /* tick line on xAxis */
  .visx-axis-top line,
  .visx-axis-bottom line {
    stroke: #e6d9ce;
    stroke-width: 1px;
  }
`;

export const socialAxisStyles = css`
  ${baseAxisStyles}

  text {
    fill: #8e9095;
    font-size: 20px;
  }

  line {
    stroke: #47494d;
    stroke-width: 1px;
  }

  .visx-axis-banding rect {
    fill: #ffffff;
    opacity: 0.4;
  }

  /* tick text on yAxis */
  .visx-axis-left text,
  .visx-axis-right text {
    fill: #8e9095;
  }

  .visx-axis-left text {
    text-anchor: end;
  }

  .visx-axis-right text {
    text-anchor: start;
  }

  /* tick line on yAxis */
  .visx-axis-left line,
  .visx-axis-right line {
    stroke: #47494d;
    stroke-width: 1px;
  }

  /* tick text on xAxis */
  .visx-axis-top text,
  .visx-axis-bottom text {
    fill: #8e9095;
    text-anchor: middle;
  }

  /* tick line on xAxis */
  .visx-axis-top line,
  .visx-axis-bottom line {
    stroke: #47494d;
    stroke-width: 1px;
  }
`;

export const videoAxisStyles = css`
  ${baseAxisStyles}

  text {
    fill: #8e9095;
    font-size: 48px;
  }

  line {
    stroke: #47494d;
    stroke-width: 3px;
  }

  .visx-axis-banding rect {
    fill: #ffffff;
    opacity: 0.4;
  }

  /* tick text on yAxis */
  .visx-axis-left text,
  .visx-axis-right text {
    fill: #8e9095;
  }

  .visx-axis-left text {
    text-anchor: end;
  }

  .visx-axis-right text {
    text-anchor: start;
  }

  /* tick line on yAxis */
  .visx-axis-left line,
  .visx-axis-right line {
    stroke: #47494d;
    stroke-width: 3px;
  }

  /* tick text on xAxis */
  .visx-axis-top text,
  .visx-axis-bottom text {
    fill: #8e9095;
    text-anchor: middle;
  }

  /* tick line on xAxis */
  .visx-axis-top line,
  .visx-axis-bottom line {
    stroke: #47494d;
    stroke-width: 3px;
  }
`;

export const printAxisStyles = css`
  ${baseAxisStyles}

  text {
    font-family: metric;
    font-size: 9.6px;
    fill: #000000;
    text-anchor: end;
  }

  line {
    stroke: #000000;
    stroke-width: 0.3px;
  }

  .visx-axis-banding rect {
    fill: #95bfc5;
    opacity: 0.4;
  }

  /* tick text on yAxis */
  .visx-axis-left text {
    text-anchor: end;
  }

  .visx-axis-right text {
    text-anchor: start;
  }

  /* tick line on yAxis */
  .visx-axis-left line,
  .visx-axis-right line {
    stroke: #000000;
    stroke-width: 0.3px;
  }

  /* tick text on xAxis */
  .visx-axis-top text,
  .visx-axis-bottom text {
    text-anchor: middle;
  }

  /* tick line on xAxis */
  .visx-axis-top line,
  .visx-axis-bottom line {
    stroke: #000000;
    stroke-width: 0.3px;
  }
`;

export const combineStyles = (initialStyles, additionalStyles) =>
  css`
    ${initialStyles}
    ${additionalStyles}
  `;

export default webAxisStyles;
