// @financial-times/vvc/src/components/axis/tick.js
/**
 * @file
 * Individual tick component
 *
 * This is needed to override @visx/axis doing stupid stuff with
 * child SVG elements.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Group } from '@visx/group';
import { Line } from '@visx/shape';

const Tick = ({ tick, labelSize, rotateDeg, axis }) => {
  const tickX = tick.to.x;
  const tickY = tick.to.y + labelSize + axis.tickLength;
  return (
    <Group className="vv-axis-tick">
      <Line from={tick.from} to={tick.to} stroke="black" />
      <text
        transform={`translate(${tickX}, ${tickY}) rotate(${rotateDeg})`}
        fontSize={labelSize}
        textAnchor="middle"
      >
        {tick.formattedValue}
      </text>
    </Group>
  );
};

Tick.defaultProps = {
  labelSize: 10,
  rotateDeg: 0,
};
Tick.propTypes = {
  tick: PropTypes.shape({
    to: PropTypes.shape({
      x: PropTypes.number,
      y: PropTypes.number,
    }),
    from: PropTypes.shape({
      x: PropTypes.number,
      y: PropTypes.number,
    }),
    formattedValue: PropTypes.string,
  }).isRequired,
  axis: PropTypes.shape({
    tickLength: PropTypes.number,
  }).isRequired,
  labelSize: PropTypes.number,
  rotateDeg: PropTypes.number,
};

export default Tick;
