// @financial-times/vvc/src/util/index.js
/**
 * @file
 * Utility functions
 */

export const convertMMtoPX = (x) => x * 2.83464480558843;

export const getResponsiveValues = (width, { topper = true, footer = true } = {}) => {
  // Web extra large
  if (width >= 1220) {
    return {
      height: 700,
      margin: { top: topper ? 100 : 0, right: 5, bottom: footer ? 60 : 0, left: 20 },
      sourceFontSize: 14,
      sourceLineHeight: 16,
      subtitleFontSize: 18,
      subtitleLineHeight: 20,
      subtitlePosition: { x: 1, y: 68 },
      tickLabelStyle: {
        fontSize: 18,
      },
      tickLength: 12,
      titleFontSize: 28,
      titleLineHeight: 28,
      titlePosition: { x: 1, y: 36 },
      axisLabelStyle: {
        fontSize: 18,
      },
      width,
    };
  }
  // Web large
  if (width >= 980) {
    return {
      height: 500,
      margin: { top: topper ? 100 : 0, right: 5, bottom: footer ? 60 : 0, left: 20 },
      sourceFontSize: 14,
      sourceLineHeight: 16,
      subtitleFontSize: 18,
      subtitleLineHeight: 20,
      subtitlePosition: { x: 1, y: 68 },
      tickLabelStyle: {
        fontSize: 18,
      },
      tickLength: 12,
      titleFontSize: 28,
      titleLineHeight: 28,
      titlePosition: { x: 1, y: 36 },
      axisLabelStyle: {
        fontSize: 18,
      },
      width,
    };
  }
  // Web medium
  if (width >= 490) {
    return {
      height: 400,
      margin: { top: topper ? 90 : 0, right: 5, bottom: footer ? 50 : 0, left: 20 },
      sourceFontSize: 14,
      sourceLineHeight: 16,
      subtitleFontSize: 18,
      subtitleLineHeight: 20,
      subtitlePosition: { x: 1, y: 64 },
      tickLabelStyle: {
        fontSize: 16,
      },
      tickLength: 12,
      titleFontSize: 24,
      titleLineHeight: 24,
      titlePosition: { x: 1, y: 32 },
      axisLabelStyle: {
        fontSize: 16,
      },
      width,
    };
  }
  // Web small
  return {
    height: 400,
    margin: { top: topper ? 90 : 0, right: 5, bottom: footer ? 50 : 0, left: 15 },
    sourceFontSize: 12,
    sourceLineHeight: 14,
    subtitleFontSize: 18,
    subtitleLineHeight: 20,
    subtitlePosition: { x: 1, y: 60 },
    tickLength: 8,
    tickLabelStyle: {
      fontSize: 14,
    },
    titleFontSize: 20,
    titleLineHeight: 20,
    titlePosition: { x: 1, y: 28 },
    axisLabelStyle: {
      fontSize: 16,
    },
    width,
  };
};

export const identity = (d) => d;

export const isZeroOrDefined = (value) => value === 0 || value;

export const stripHTMLTags = (string) =>
  string.replace(/(<([^>]+)>)/gi, '').replace(/(&nbsp;)/gi, ' ');
