// @financial-times/vvc/src/components/voronoi/region.js
/**
 * @file
 * Single Voronoi region
 */

import React, { useState } from 'react';
import { VoronoiPolygon } from '@visx/voronoi';
import { Circle, LinePath } from '@visx/shape';
import PropTypes from 'prop-types';

const Region = ({
  x,
  y,
  polygon,
  debug,
  onMouseOver,
  onFocus,
  onMouseOut,
  onBlur,
  fill,
  circleStroke,
  circleStrokeWidth,
  r,
  showLine,
  height,
  showCircle,
}) => {
  const [isActive, setIsActive] = useState(false);
  const setActive = () => setIsActive(true);
  const setInactive = () => setIsActive(false);
  return (
    <g onMouseOver={setActive} onMouseOut={setInactive} onFocus={setActive} onBlur={setInactive}>
      {showCircle && isActive && showLine && (
        <LinePath
          stroke="#33302e"
          data={[
            { x, y: 0 },
            { x, y: height },
          ]}
          x={({ x }) => x}
          y={({ y }) => y}
          strokeDasharray="3, 1"
        />
      )}
      {showCircle && isActive && (
        <Circle
          fill={fill}
          r={r}
          cx={x}
          cy={y}
          stroke={circleStroke}
          strokeWidth={circleStrokeWidth}
        />
      )}
      <VoronoiPolygon
        polygon={polygon}
        strokeWidth={debug ? 1 : 0}
        stroke="red"
        fill="transparent"
        onMouseOver={onMouseOver}
        onFocus={onFocus}
        onMouseOut={onMouseOut}
        onBlur={onBlur}
      />
    </g>
  );
};

Region.propTypes = {
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
  polygon: PropTypes.any.isRequired,
  debug: PropTypes.bool,
  onMouseOver: PropTypes.func.isRequired,
  onFocus: PropTypes.func.isRequired,
  fill: PropTypes.string,
  r: PropTypes.number,
  circleStroke: PropTypes.string,
  circleStrokeWidth: PropTypes.number,
  showLine: PropTypes.bool,
  height: PropTypes.number,
};

Region.defaultProps = {
  fill: 'black',
  r: 4,
  debug: false,
  circleStroke: 'transparent',
  circleStrokeWidth: 1,
  showLine: false,
  height: 1,
};

export default Region;
