// @financial-times/vvc/src/sample-data/date-continuous.js
const dateContinuousSampleData = [
  {
    date: '05/04/2007',
    Pearson: 850,
    Sky: 562,
    Bloomsbury: 203.5,
    'Vitesse Media': 32.5,
    Totally: 286.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/04/2007',
    Pearson: 846,
    Sky: 571.5,
    Bloomsbury: 200,
    'Vitesse Media': 32.5,
    Totally: 286.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/04/2007',
    Pearson: 856,
    Sky: 572.5,
    Bloomsbury: 204,
    'Vitesse Media': 33,
    Totally: 273.81,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/04/2007',
    Pearson: 863.5,
    Sky: 568.5,
    Bloomsbury: 191.25,
    'Vitesse Media': 33.5,
    Totally: 261.37,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/05/2007',
    Pearson: 875.5,
    Sky: 628.5,
    Bloomsbury: 183.5,
    'Vitesse Media': 33.5,
    Totally: 273.81,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/05/2007',
    Pearson: 875.5,
    Sky: 621.5,
    Bloomsbury: 189,
    'Vitesse Media': 34,
    Totally: 261.37,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/05/2007',
    Pearson: 901.5,
    Sky: 641.5,
    Bloomsbury: 183,
    'Vitesse Media': 34,
    Totally: 261.37,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/05/2007',
    Pearson: 882,
    Sky: 634.5,
    Bloomsbury: 185,
    'Vitesse Media': 35,
    Totally: 261.37,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '31/05/2007',
    Pearson: 898,
    Sky: 660,
    Bloomsbury: 186.25,
    'Vitesse Media': 35,
    Totally: 261.37,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/06/2007',
    Pearson: 851.5,
    Sky: 640,
    Bloomsbury: 182.5,
    'Vitesse Media': 35,
    Totally: 273.81,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/06/2007',
    Pearson: 863,
    Sky: 646,
    Bloomsbury: 191,
    'Vitesse Media': 35,
    Totally: 286.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/06/2007',
    Pearson: 824.5,
    Sky: 634.5,
    Bloomsbury: 190,
    'Vitesse Media': 35,
    Totally: 286.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/06/2007',
    Pearson: 835,
    Sky: 631,
    Bloomsbury: 194,
    'Vitesse Media': 35,
    Totally: 286.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/07/2007',
    Pearson: 827.5,
    Sky: 649,
    Bloomsbury: 193.25,
    'Vitesse Media': 37.5,
    Totally: 286.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/07/2007',
    Pearson: 825.5,
    Sky: 709,
    Bloomsbury: 185.25,
    'Vitesse Media': 40.5,
    Totally: 286.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/07/2007',
    Pearson: 824.5,
    Sky: 705.5,
    Bloomsbury: 184,
    'Vitesse Media': 39,
    Totally: 286.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/07/2007',
    Pearson: 769.5,
    Sky: 675.5,
    Bloomsbury: 176,
    'Vitesse Media': 34.5,
    Totally: 286.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/08/2007',
    Pearson: 797,
    Sky: 660,
    Bloomsbury: 154,
    'Vitesse Media': 34.5,
    Totally: 286.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/08/2007',
    Pearson: 774,
    Sky: 694.5,
    Bloomsbury: 168.25,
    'Vitesse Media': 34.5,
    Totally: 248.92,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/08/2007',
    Pearson: 737,
    Sky: 646.5,
    Bloomsbury: 162,
    'Vitesse Media': 33.5,
    Totally: 248.92,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/08/2007',
    Pearson: 747,
    Sky: 658,
    Bloomsbury: 164,
    'Vitesse Media': 36,
    Totally: 224.03,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/08/2007',
    Pearson: 746.5,
    Sky: 667,
    Bloomsbury: 163,
    'Vitesse Media': 36,
    Totally: 224.03,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/09/2007',
    Pearson: 744,
    Sky: 674,
    Bloomsbury: 159,
    'Vitesse Media': 36,
    Totally: 211.58,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/09/2007',
    Pearson: 747.5,
    Sky: 677.5,
    Bloomsbury: 165,
    'Vitesse Media': 36,
    Totally: 211.58,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/09/2007',
    Pearson: 759.5,
    Sky: 679,
    Bloomsbury: 158.5,
    'Vitesse Media': 36,
    Totally: 211.58,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/09/2007',
    Pearson: 754.5,
    Sky: 694,
    Bloomsbury: 153,
    'Vitesse Media': 36,
    Totally: 211.58,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/10/2007',
    Pearson: 751.5,
    Sky: 693,
    Bloomsbury: 144,
    'Vitesse Media': 35.5,
    Totally: 211.58,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/10/2007',
    Pearson: 791,
    Sky: 696,
    Bloomsbury: 145.5,
    'Vitesse Media': 35.5,
    Totally: 211.58,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/10/2007',
    Pearson: 766,
    Sky: 676,
    Bloomsbury: 155,
    'Vitesse Media': 35.5,
    Totally: 224.03,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/10/2007',
    Pearson: 784,
    Sky: 655.5,
    Bloomsbury: 156.75,
    'Vitesse Media': 35,
    Totally: 199.14,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/11/2007',
    Pearson: 790.5,
    Sky: 690,
    Bloomsbury: 168.25,
    'Vitesse Media': 38,
    Totally: 199.14,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/11/2007',
    Pearson: 753.5,
    Sky: 617.5,
    Bloomsbury: 151.75,
    'Vitesse Media': 38,
    Totally: 199.14,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/11/2007',
    Pearson: 738.5,
    Sky: 626,
    Bloomsbury: 136.5,
    'Vitesse Media': 38,
    Totally: 199.14,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/11/2007',
    Pearson: 714,
    Sky: 623,
    Bloomsbury: 119,
    'Vitesse Media': 38,
    Totally: 174.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/11/2007',
    Pearson: 749.5,
    Sky: 634.5,
    Bloomsbury: 130,
    'Vitesse Media': 37,
    Totally: 161.8,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/12/2007',
    Pearson: 735,
    Sky: 604.5,
    Bloomsbury: 122.25,
    'Vitesse Media': 37,
    Totally: 174.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/12/2007',
    Pearson: 704,
    Sky: 599,
    Bloomsbury: 129.5,
    'Vitesse Media': 37,
    Totally: 174.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/12/2007',
    Pearson: 711.5,
    Sky: 600,
    Bloomsbury: 144.5,
    'Vitesse Media': 37,
    Totally: 174.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/12/2007',
    Pearson: 742.5,
    Sky: 618,
    Bloomsbury: 153,
    'Vitesse Media': 37,
    Totally: 174.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/01/2008',
    Pearson: 720,
    Sky: 599.5,
    Bloomsbury: 150,
    'Vitesse Media': 37,
    Totally: 149.35,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/01/2008',
    Pearson: 672,
    Sky: 549,
    Bloomsbury: 152,
    'Vitesse Media': 37,
    Totally: 149.35,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/01/2008',
    Pearson: 649,
    Sky: 521,
    Bloomsbury: 161,
    'Vitesse Media': 37,
    Totally: 149.35,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/01/2008',
    Pearson: 669,
    Sky: 551,
    Bloomsbury: 161,
    'Vitesse Media': 37,
    Totally: 136.91,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '31/01/2008',
    Pearson: 690.5,
    Sky: 550,
    Bloomsbury: 157.75,
    'Vitesse Media': 37,
    Totally: 136.91,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/02/2008',
    Pearson: 641,
    Sky: 558,
    Bloomsbury: 153,
    'Vitesse Media': 32.5,
    Totally: 124.46,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/02/2008',
    Pearson: 674,
    Sky: 588,
    Bloomsbury: 156,
    'Vitesse Media': 30.5,
    Totally: 112.02,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/02/2008',
    Pearson: 681,
    Sky: 595,
    Bloomsbury: 156,
    'Vitesse Media': 30.5,
    Totally: 124.46,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/02/2008',
    Pearson: 680,
    Sky: 571,
    Bloomsbury: 172,
    'Vitesse Media': 30.5,
    Totally: 136.91,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/03/2008',
    Pearson: 688.5,
    Sky: 551,
    Bloomsbury: 162,
    'Vitesse Media': 30.5,
    Totally: 136.91,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/03/2008',
    Pearson: 678,
    Sky: 547,
    Bloomsbury: 159.5,
    'Vitesse Media': 28.5,
    Totally: 136.91,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/03/2008',
    Pearson: 682,
    Sky: 537.5,
    Bloomsbury: 146,
    'Vitesse Media': 27.75,
    Totally: 112.02,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/03/2008',
    Pearson: 684,
    Sky: 566,
    Bloomsbury: 148,
    'Vitesse Media': 26,
    Totally: 112.02,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/04/2008',
    Pearson: 698,
    Sky: 569,
    Bloomsbury: 160,
    'Vitesse Media': 25.5,
    Totally: 112.02,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/04/2008',
    Pearson: 668.5,
    Sky: 558,
    Bloomsbury: 152,
    'Vitesse Media': 25.5,
    Totally: 99.57,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/04/2008',
    Pearson: 652.5,
    Sky: 544,
    Bloomsbury: 133.5,
    'Vitesse Media': 25.5,
    Totally: 99.57,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/04/2008',
    Pearson: 665.5,
    Sky: 536.5,
    Bloomsbury: 128,
    'Vitesse Media': 25.5,
    Totally: 112.02,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/05/2008',
    Pearson: 660.5,
    Sky: 548,
    Bloomsbury: 140.5,
    'Vitesse Media': 25.5,
    Totally: 112.02,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/05/2008',
    Pearson: 698.5,
    Sky: 530,
    Bloomsbury: 147,
    'Vitesse Media': 25.5,
    Totally: 112.02,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/05/2008',
    Pearson: 688,
    Sky: 516.5,
    Bloomsbury: 147,
    'Vitesse Media': 25,
    Totally: 124.46,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/05/2008',
    Pearson: 672,
    Sky: 531.5,
    Bloomsbury: 150,
    'Vitesse Media': 25,
    Totally: 124.46,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/05/2008',
    Pearson: 666,
    Sky: 539.5,
    Bloomsbury: 135,
    'Vitesse Media': 25,
    Totally: 124.46,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/06/2008',
    Pearson: 680,
    Sky: 544,
    Bloomsbury: 139.25,
    'Vitesse Media': 25,
    Totally: 112.02,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/06/2008',
    Pearson: 648.5,
    Sky: 508.5,
    Bloomsbury: 125.5,
    'Vitesse Media': 25,
    Totally: 112.02,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/06/2008',
    Pearson: 633.5,
    Sky: 504.5,
    Bloomsbury: 129,
    'Vitesse Media': 23.5,
    Totally: 124.46,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/06/2008',
    Pearson: 610.5,
    Sky: 470.75,
    Bloomsbury: 141.5,
    'Vitesse Media': 23.5,
    Totally: 124.46,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/07/2008',
    Pearson: 597.5,
    Sky: 458,
    Bloomsbury: 139.5,
    'Vitesse Media': 23.5,
    Totally: 99.57,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/07/2008',
    Pearson: 611.5,
    Sky: 443.25,
    Bloomsbury: 150.75,
    'Vitesse Media': 23.5,
    Totally: 99.57,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/07/2008',
    Pearson: 606,
    Sky: 451.75,
    Bloomsbury: 151,
    'Vitesse Media': 23.25,
    Totally: 87.12,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/07/2008',
    Pearson: 595.5,
    Sky: 466.5,
    Bloomsbury: 144,
    'Vitesse Media': 23.25,
    Totally: 99.57,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '31/07/2008',
    Pearson: 650.5,
    Sky: 454.25,
    Bloomsbury: 154,
    'Vitesse Media': 23.25,
    Totally: 87.12,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/08/2008',
    Pearson: 649.5,
    Sky: 462.75,
    Bloomsbury: 162,
    'Vitesse Media': 23.25,
    Totally: 87.12,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/08/2008',
    Pearson: 668.5,
    Sky: 481.5,
    Bloomsbury: 157,
    'Vitesse Media': 23.25,
    Totally: 87.12,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/08/2008',
    Pearson: 643.5,
    Sky: 441.75,
    Bloomsbury: 164.5,
    'Vitesse Media': 20.25,
    Totally: 87.12,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/08/2008',
    Pearson: 677.5,
    Sky: 460.75,
    Bloomsbury: 165.75,
    'Vitesse Media': 20.25,
    Totally: 99.57,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/09/2008',
    Pearson: 675,
    Sky: 460,
    Bloomsbury: 167,
    'Vitesse Media': 20.25,
    Totally: 99.57,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/09/2008',
    Pearson: 674.5,
    Sky: 445.5,
    Bloomsbury: 171.75,
    'Vitesse Media': 20.25,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/09/2008',
    Pearson: 594,
    Sky: 419,
    Bloomsbury: 161,
    'Vitesse Media': 20.75,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/09/2008',
    Pearson: 605,
    Sky: 431.75,
    Bloomsbury: 167.75,
    'Vitesse Media': 20.75,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/10/2008',
    Pearson: 624.5,
    Sky: 429.5,
    Bloomsbury: 155,
    'Vitesse Media': 20,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/10/2008',
    Pearson: 543.5,
    Sky: 400,
    Bloomsbury: 168.25,
    'Vitesse Media': 19.5,
    Totally: 44.81,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/10/2008',
    Pearson: 533.5,
    Sky: 361.5,
    Bloomsbury: 163,
    'Vitesse Media': 19.5,
    Totally: 44.81,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/10/2008',
    Pearson: 602.5,
    Sky: 372.25,
    Bloomsbury: 159,
    'Vitesse Media': 19,
    Totally: 44.81,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/10/2008',
    Pearson: 595,
    Sky: 378,
    Bloomsbury: 151,
    'Vitesse Media': 19,
    Totally: 44.81,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/11/2008',
    Pearson: 596.5,
    Sky: 404.5,
    Bloomsbury: 162,
    'Vitesse Media': 19,
    Totally: 44.81,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/11/2008',
    Pearson: 582,
    Sky: 394.75,
    Bloomsbury: 160.5,
    'Vitesse Media': 19,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/11/2008',
    Pearson: 587,
    Sky: 375,
    Bloomsbury: 165,
    'Vitesse Media': 19,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/11/2008',
    Pearson: 602.5,
    Sky: 414.25,
    Bloomsbury: 157,
    'Vitesse Media': 19,
    Totally: 27.38,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/12/2008',
    Pearson: 624.5,
    Sky: 438.75,
    Bloomsbury: 152,
    'Vitesse Media': 19,
    Totally: 27.38,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/12/2008',
    Pearson: 632,
    Sky: 471,
    Bloomsbury: 137.5,
    'Vitesse Media': 19,
    Totally: 27.38,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/12/2008',
    Pearson: 650.5,
    Sky: 478.75,
    Bloomsbury: 150.5,
    'Vitesse Media': 19,
    Totally: 27.38,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/12/2008',
    Pearson: 643,
    Sky: 468.25,
    Bloomsbury: 153.5,
    'Vitesse Media': 19,
    Totally: 27.38,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/01/2009',
    Pearson: 641,
    Sky: 480,
    Bloomsbury: 159.75,
    'Vitesse Media': 18.5,
    Totally: 29.87,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/01/2009',
    Pearson: 612,
    Sky: 475,
    Bloomsbury: 153.75,
    'Vitesse Media': 18.5,
    Totally: 29.87,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/01/2009',
    Pearson: 588.5,
    Sky: 405.75,
    Bloomsbury: 157.5,
    'Vitesse Media': 18.5,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/01/2009',
    Pearson: 643,
    Sky: 424.25,
    Bloomsbury: 147,
    'Vitesse Media': 18.5,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/01/2009',
    Pearson: 668.5,
    Sky: 500,
    Bloomsbury: 144,
    'Vitesse Media': 16,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/02/2009',
    Pearson: 672,
    Sky: 486.5,
    Bloomsbury: 141.5,
    'Vitesse Media': 16,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/02/2009',
    Pearson: 655.5,
    Sky: 465.25,
    Bloomsbury: 146,
    'Vitesse Media': 16,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/02/2009',
    Pearson: 641,
    Sky: 458,
    Bloomsbury: 135,
    'Vitesse Media': 16,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/02/2009',
    Pearson: 654.5,
    Sky: 470.5,
    Bloomsbury: 127,
    'Vitesse Media': 16,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/03/2009',
    Pearson: 654.5,
    Sky: 448.5,
    Bloomsbury: 140.5,
    'Vitesse Media': 16,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/03/2009',
    Pearson: 680.5,
    Sky: 446.25,
    Bloomsbury: 137,
    'Vitesse Media': 16,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/03/2009',
    Pearson: 693.5,
    Sky: 458.5,
    Bloomsbury: 134.5,
    'Vitesse Media': 16,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/03/2009',
    Pearson: 697,
    Sky: 435.5,
    Bloomsbury: 135,
    'Vitesse Media': 16,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/04/2009',
    Pearson: 732.5,
    Sky: 432.25,
    Bloomsbury: 119.5,
    'Vitesse Media': 16,
    Totally: 32.36,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/04/2009',
    Pearson: 687.5,
    Sky: 425.75,
    Bloomsbury: 122.25,
    'Vitesse Media': 16,
    Totally: 32.36,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/04/2009',
    Pearson: 670,
    Sky: 440,
    Bloomsbury: 127.5,
    'Vitesse Media': 16,
    Totally: 32.36,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/04/2009',
    Pearson: 674,
    Sky: 440.25,
    Bloomsbury: 125,
    'Vitesse Media': 16,
    Totally: 32.36,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/04/2009',
    Pearson: 706.5,
    Sky: 486.75,
    Bloomsbury: 125,
    'Vitesse Media': 16,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/05/2009',
    Pearson: 719,
    Sky: 481.25,
    Bloomsbury: 121,
    'Vitesse Media': 16,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/05/2009',
    Pearson: 693,
    Sky: 468,
    Bloomsbury: 120,
    'Vitesse Media': 14.5,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/05/2009',
    Pearson: 671,
    Sky: 450.75,
    Bloomsbury: 127.25,
    'Vitesse Media': 14.5,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/05/2009',
    Pearson: 666.5,
    Sky: 452.5,
    Bloomsbury: 120.5,
    'Vitesse Media': 14.5,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/06/2009',
    Pearson: 631.5,
    Sky: 452.75,
    Bloomsbury: 123.5,
    'Vitesse Media': 14.5,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/06/2009',
    Pearson: 617,
    Sky: 449,
    Bloomsbury: 122.5,
    'Vitesse Media': 14.5,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/06/2009',
    Pearson: 604,
    Sky: 431,
    Bloomsbury: 118,
    'Vitesse Media': 14,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/06/2009',
    Pearson: 614.5,
    Sky: 451.75,
    Bloomsbury: 123.5,
    'Vitesse Media': 14,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/07/2009',
    Pearson: 585.5,
    Sky: 454.75,
    Bloomsbury: 125,
    'Vitesse Media': 13.5,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/07/2009',
    Pearson: 586.5,
    Sky: 477.75,
    Bloomsbury: 128.5,
    'Vitesse Media': 13.5,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/07/2009',
    Pearson: 611.5,
    Sky: 490.25,
    Bloomsbury: 136.25,
    'Vitesse Media': 13.5,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/07/2009',
    Pearson: 623,
    Sky: 517,
    Bloomsbury: 142.5,
    'Vitesse Media': 13.5,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/07/2009',
    Pearson: 684.5,
    Sky: 546,
    Bloomsbury: 139,
    'Vitesse Media': 13.5,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/08/2009',
    Pearson: 704.5,
    Sky: 534,
    Bloomsbury: 136.5,
    'Vitesse Media': 13.5,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/08/2009',
    Pearson: 721,
    Sky: 533.5,
    Bloomsbury: 131,
    'Vitesse Media': 13.5,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/08/2009',
    Pearson: 722.5,
    Sky: 542.5,
    Bloomsbury: 130,
    'Vitesse Media': 13.5,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/08/2009',
    Pearson: 745,
    Sky: 540.5,
    Bloomsbury: 129.75,
    'Vitesse Media': 13.5,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/09/2009',
    Pearson: 735,
    Sky: 535.5,
    Bloomsbury: 135,
    'Vitesse Media': 13.5,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/09/2009',
    Pearson: 744,
    Sky: 541,
    Bloomsbury: 127,
    'Vitesse Media': 13.5,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/09/2009',
    Pearson: 754.5,
    Sky: 541.5,
    Bloomsbury: 125.25,
    'Vitesse Media': 13.5,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/09/2009',
    Pearson: 759,
    Sky: 560,
    Bloomsbury: 120.5,
    'Vitesse Media': 13.5,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/10/2009',
    Pearson: 767.5,
    Sky: 573,
    Bloomsbury: 123.5,
    'Vitesse Media': 13.5,
    Totally: 67.21,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/10/2009',
    Pearson: 786,
    Sky: 581,
    Bloomsbury: 129.5,
    'Vitesse Media': 12,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/10/2009',
    Pearson: 808,
    Sky: 571.5,
    Bloomsbury: 131,
    'Vitesse Media': 12.25,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/10/2009',
    Pearson: 831,
    Sky: 559.5,
    Bloomsbury: 130,
    'Vitesse Media': 12.25,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/10/2009',
    Pearson: 841,
    Sky: 539,
    Bloomsbury: 128,
    'Vitesse Media': 12.25,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/11/2009',
    Pearson: 843,
    Sky: 542.5,
    Bloomsbury: 132,
    'Vitesse Media': 12.25,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/11/2009',
    Pearson: 845.5,
    Sky: 539.5,
    Bloomsbury: 126.25,
    'Vitesse Media': 12.25,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/11/2009',
    Pearson: 837,
    Sky: 540,
    Bloomsbury: 125,
    'Vitesse Media': 12.25,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/11/2009',
    Pearson: 825,
    Sky: 532.5,
    Bloomsbury: 127,
    'Vitesse Media': 12.25,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/12/2009',
    Pearson: 848,
    Sky: 543.5,
    Bloomsbury: 126.25,
    'Vitesse Media': 12.25,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/12/2009',
    Pearson: 862,
    Sky: 564.5,
    Bloomsbury: 127.5,
    'Vitesse Media': 12.25,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/12/2009',
    Pearson: 863.5,
    Sky: 565,
    Bloomsbury: 125,
    'Vitesse Media': 12.25,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/12/2009',
    Pearson: 878,
    Sky: 561.5,
    Bloomsbury: 129.5,
    'Vitesse Media': 12.25,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '31/12/2009',
    Pearson: 891,
    Sky: 562,
    Bloomsbury: 126.5,
    'Vitesse Media': 12.25,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/01/2010',
    Pearson: 869.5,
    Sky: 568.5,
    Bloomsbury: 126.5,
    'Vitesse Media': 12.25,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/01/2010',
    Pearson: 887.5,
    Sky: 572.5,
    Bloomsbury: 130,
    'Vitesse Media': 12.25,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/01/2010',
    Pearson: 891.5,
    Sky: 560,
    Bloomsbury: 130,
    'Vitesse Media': 11,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/01/2010',
    Pearson: 874.5,
    Sky: 540,
    Bloomsbury: 130,
    'Vitesse Media': 11,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/02/2010',
    Pearson: 870.5,
    Sky: 532.5,
    Bloomsbury: 128,
    'Vitesse Media': 10.5,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/02/2010',
    Pearson: 865,
    Sky: 528,
    Bloomsbury: 124,
    'Vitesse Media': 10.5,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/02/2010',
    Pearson: 875.5,
    Sky: 534.5,
    Bloomsbury: 125,
    'Vitesse Media': 10.5,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/02/2010',
    Pearson: 898.5,
    Sky: 538,
    Bloomsbury: 125.25,
    'Vitesse Media': 11.5,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/03/2010',
    Pearson: 974,
    Sky: 558.5,
    Bloomsbury: 122.5,
    'Vitesse Media': 11.5,
    Totally: 29.87,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/03/2010',
    Pearson: 994,
    Sky: 569.5,
    Bloomsbury: 122.5,
    'Vitesse Media': 11.5,
    Totally: 29.87,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/03/2010',
    Pearson: 1003,
    Sky: 588,
    Bloomsbury: 118,
    'Vitesse Media': 11.5,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/03/2010',
    Pearson: 1036,
    Sky: 600,
    Bloomsbury: 119,
    'Vitesse Media': 10.5,
    Totally: 47.3,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/04/2010',
    Pearson: 1044,
    Sky: 612.5,
    Bloomsbury: 115,
    'Vitesse Media': 10.5,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/04/2010',
    Pearson: 997.5,
    Sky: 602.5,
    Bloomsbury: 115.5,
    'Vitesse Media': 10,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/04/2010',
    Pearson: 1010,
    Sky: 623.5,
    Bloomsbury: 112,
    'Vitesse Media': 10,
    Totally: 49.78,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/04/2010',
    Pearson: 1019,
    Sky: 628,
    Bloomsbury: 110,
    'Vitesse Media': 10,
    Totally: 44.81,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/04/2010',
    Pearson: 1050,
    Sky: 625,
    Bloomsbury: 111.5,
    'Vitesse Media': 10.5,
    Totally: 72.19,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/05/2010',
    Pearson: 979,
    Sky: 600,
    Bloomsbury: 112.5,
    'Vitesse Media': 10.5,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/05/2010',
    Pearson: 1001,
    Sky: 597,
    Bloomsbury: 108.25,
    'Vitesse Media': 10.5,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/05/2010',
    Pearson: 943,
    Sky: 560,
    Bloomsbury: 106.5,
    'Vitesse Media': 10.5,
    Totally: 64.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/05/2010',
    Pearson: 948,
    Sky: 575.5,
    Bloomsbury: 109.5,
    'Vitesse Media': 9,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/06/2010',
    Pearson: 958.5,
    Sky: 580,
    Bloomsbury: 108,
    'Vitesse Media': 9,
    Totally: 52.27,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/06/2010',
    Pearson: 938,
    Sky: 585.5,
    Bloomsbury: 116.5,
    'Vitesse Media': 9,
    Totally: 49.78,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/06/2010',
    Pearson: 956.5,
    Sky: 699,
    Bloomsbury: 113.5,
    'Vitesse Media': 9,
    Totally: 47.3,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/06/2010',
    Pearson: 927,
    Sky: 702,
    Bloomsbury: 112,
    'Vitesse Media': 9,
    Totally: 49.78,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/07/2010',
    Pearson: 864,
    Sky: 695.5,
    Bloomsbury: 115,
    'Vitesse Media': 9,
    Totally: 44.81,
    highlight: '',
    annotate: 'annotation with elbow line',
    type: 'elbow',
  },
  {
    date: '08/07/2010',
    Pearson: 907,
    Sky: 694.5,
    Bloomsbury: 114,
    'Vitesse Media': 9,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/07/2010',
    Pearson: 936.5,
    Sky: 699.5,
    Bloomsbury: 116,
    'Vitesse Media': 9,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/07/2010',
    Pearson: 962,
    Sky: 700,
    Bloomsbury: 114,
    'Vitesse Media': 9,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/07/2010',
    Pearson: 1012,
    Sky: 711,
    Bloomsbury: 115.75,
    'Vitesse Media': 9,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/08/2010',
    Pearson: 1020,
    Sky: 708.5,
    Bloomsbury: 126.5,
    'Vitesse Media': 9,
    Totally: 45.8,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/08/2010',
    Pearson: 983.5,
    Sky: 706,
    Bloomsbury: 118,
    'Vitesse Media': 9,
    Totally: 47.3,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/08/2010',
    Pearson: 952,
    Sky: 706,
    Bloomsbury: 111.25,
    'Vitesse Media': 9.25,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/08/2010',
    Pearson: 948.5,
    Sky: 706,
    Bloomsbury: 112,
    'Vitesse Media': 9.25,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/09/2010',
    Pearson: 996.5,
    Sky: 710.5,
    Bloomsbury: 110,
    'Vitesse Media': 9.25,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/09/2010',
    Pearson: 1010,
    Sky: 717,
    Bloomsbury: 115,
    'Vitesse Media': 9.25,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/09/2010',
    Pearson: 1011,
    Sky: 709.5,
    Bloomsbury: 119.75,
    'Vitesse Media': 9,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/09/2010',
    Pearson: 994.5,
    Sky: 708,
    Bloomsbury: 115.5,
    'Vitesse Media': 9,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/09/2010',
    Pearson: 985.5,
    Sky: 705.5,
    Bloomsbury: 117.75,
    'Vitesse Media': 9,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/10/2010',
    Pearson: 991.5,
    Sky: 707.5,
    Bloomsbury: 124.5,
    'Vitesse Media': 9,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/10/2010',
    Pearson: 991,
    Sky: 707,
    Bloomsbury: 125,
    'Vitesse Media': 9.25,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/10/2010',
    Pearson: 986,
    Sky: 697.5,
    Bloomsbury: 125,
    'Vitesse Media': 9.25,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/10/2010',
    Pearson: 954.5,
    Sky: 706,
    Bloomsbury: 126,
    'Vitesse Media': 9.25,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/11/2010',
    Pearson: 973.5,
    Sky: 711.5,
    Bloomsbury: 124,
    'Vitesse Media': 9.25,
    Totally: 47.3,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/11/2010',
    Pearson: 957,
    Sky: 726.5,
    Bloomsbury: 117,
    'Vitesse Media': 9.25,
    Totally: 60.24,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/11/2010',
    Pearson: 937.5,
    Sky: 726,
    Bloomsbury: 113.25,
    'Vitesse Media': 9.25,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/11/2010',
    Pearson: 954.5,
    Sky: 721,
    Bloomsbury: 115,
    'Vitesse Media': 9.25,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/12/2010',
    Pearson: 982,
    Sky: 729,
    Bloomsbury: 119.25,
    'Vitesse Media': 9.25,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/12/2010',
    Pearson: 1014,
    Sky: 736,
    Bloomsbury: 117,
    'Vitesse Media': 9.25,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/12/2010',
    Pearson: 1024,
    Sky: 731,
    Bloomsbury: 122,
    'Vitesse Media': 9.25,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/12/2010',
    Pearson: 1027,
    Sky: 742,
    Bloomsbury: 122.5,
    'Vitesse Media': 9.25,
    Totally: 44.81,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/12/2010',
    Pearson: 1015,
    Sky: 736,
    Bloomsbury: 122.25,
    'Vitesse Media': 9.25,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/01/2011',
    Pearson: 1013,
    Sky: 746,
    Bloomsbury: 123,
    'Vitesse Media': 9.25,
    Totally: 64.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/01/2011',
    Pearson: 1010,
    Sky: 740.5,
    Bloomsbury: 120,
    'Vitesse Media': 9,
    Totally: 64.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/01/2011',
    Pearson: 1039,
    Sky: 746,
    Bloomsbury: 116.75,
    'Vitesse Media': 9,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/01/2011',
    Pearson: 1049,
    Sky: 762.5,
    Bloomsbury: 117,
    'Vitesse Media': 7,
    Totally: 47.3,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/02/2011',
    Pearson: 1040,
    Sky: 767.5,
    Bloomsbury: 114,
    'Vitesse Media': 7,
    Totally: 72.19,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/02/2011',
    Pearson: 1060,
    Sky: 759,
    Bloomsbury: 114,
    'Vitesse Media': 7,
    Totally: 60.24,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/02/2011',
    Pearson: 1048,
    Sky: 749,
    Bloomsbury: 115,
    'Vitesse Media': 8,
    Totally: 67.21,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/02/2011',
    Pearson: 1013,
    Sky: 755.5,
    Bloomsbury: 112.5,
    'Vitesse Media': 8,
    Totally: 77.17,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/03/2011',
    Pearson: 1054,
    Sky: 823,
    Bloomsbury: 124.25,
    'Vitesse Media': 8,
    Totally: 77.17,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/03/2011',
    Pearson: 1105,
    Sky: 827,
    Bloomsbury: 120,
    'Vitesse Media': 8,
    Totally: 69.7,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/03/2011',
    Pearson: 1054,
    Sky: 821,
    Bloomsbury: 118.75,
    'Vitesse Media': 8,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/03/2011',
    Pearson: 1077,
    Sky: 828,
    Bloomsbury: 127.13,
    'Vitesse Media': 8,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '31/03/2011',
    Pearson: 1101,
    Sky: 825,
    Bloomsbury: 130,
    'Vitesse Media': 8,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/04/2011',
    Pearson: 1098,
    Sky: 833.5,
    Bloomsbury: 131.5,
    'Vitesse Media': 8,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/04/2011',
    Pearson: 1106,
    Sky: 834.5,
    Bloomsbury: 132,
    'Vitesse Media': 8,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/04/2011',
    Pearson: 1131,
    Sky: 835,
    Bloomsbury: 131,
    'Vitesse Media': 8,
    Totally: 64.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/04/2011',
    Pearson: 1151,
    Sky: 842,
    Bloomsbury: 135,
    'Vitesse Media': 8,
    Totally: 64.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/05/2011',
    Pearson: 1125,
    Sky: 839,
    Bloomsbury: 135,
    'Vitesse Media': 8,
    Totally: 64.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/05/2011',
    Pearson: 1160,
    Sky: 840,
    Bloomsbury: 135.5,
    'Vitesse Media': 8,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/05/2011',
    Pearson: 1158,
    Sky: 842.5,
    Bloomsbury: 132,
    'Vitesse Media': 8,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/05/2011',
    Pearson: 1121,
    Sky: 838,
    Bloomsbury: 132,
    'Vitesse Media': 8,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/06/2011',
    Pearson: 1135,
    Sky: 828.5,
    Bloomsbury: 128,
    'Vitesse Media': 8,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/06/2011',
    Pearson: 1161,
    Sky: 835.5,
    Bloomsbury: 130,
    'Vitesse Media': 8.5,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/06/2011',
    Pearson: 1146,
    Sky: 826.5,
    Bloomsbury: 129,
    'Vitesse Media': 8.5,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/06/2011',
    Pearson: 1133,
    Sky: 835,
    Bloomsbury: 128,
    'Vitesse Media': 8.5,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/06/2011',
    Pearson: 1176,
    Sky: 846.5,
    Bloomsbury: 129.75,
    'Vitesse Media': 8.5,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/07/2011',
    Pearson: 1202,
    Sky: 812,
    Bloomsbury: 133,
    'Vitesse Media': 8.5,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/07/2011',
    Pearson: 1156,
    Sky: 696,
    Bloomsbury: 114.37,
    'Vitesse Media': 8.5,
    Totally: 69.7,
    highlight: '',
    annotate: 'type not given',
    type: '',
  },
  {
    date: '21/07/2011',
    Pearson: 1155,
    Sky: 741,
    Bloomsbury: 114.25,
    'Vitesse Media': 8.5,
    Totally: 67.21,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/07/2011',
    Pearson: 1139,
    Sky: 716,
    Bloomsbury: 113,
    'Vitesse Media': 8.5,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/08/2011',
    Pearson: 1119,
    Sky: 659.5,
    Bloomsbury: 102,
    'Vitesse Media': 8.5,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/08/2011',
    Pearson: 1077,
    Sky: 654,
    Bloomsbury: 104.5,
    'Vitesse Media': 8.5,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/08/2011',
    Pearson: 1038,
    Sky: 634.5,
    Bloomsbury: 102.75,
    'Vitesse Media': 8.5,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/08/2011',
    Pearson: 1063,
    Sky: 622,
    Bloomsbury: 100,
    'Vitesse Media': 7.5,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/09/2011',
    Pearson: 1122,
    Sky: 673,
    Bloomsbury: 100,
    'Vitesse Media': 7.5,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/09/2011',
    Pearson: 1097,
    Sky: 669,
    Bloomsbury: 100.5,
    'Vitesse Media': 7.5,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/09/2011',
    Pearson: 1110,
    Sky: 699.5,
    Bloomsbury: 101.25,
    'Vitesse Media': 7.5,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/09/2011',
    Pearson: 1116,
    Sky: 668.5,
    Bloomsbury: 99,
    'Vitesse Media': 7,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/09/2011',
    Pearson: 1144,
    Sky: 682.5,
    Bloomsbury: 98,
    'Vitesse Media': 7,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/10/2011',
    Pearson: 1156,
    Sky: 689.5,
    Bloomsbury: 97,
    'Vitesse Media': 5.5,
    Totally: 42.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/10/2011',
    Pearson: 1162,
    Sky: 679.5,
    Bloomsbury: 98.25,
    'Vitesse Media': 4.75,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/10/2011',
    Pearson: 1135,
    Sky: 725,
    Bloomsbury: 97.75,
    'Vitesse Media': 4.75,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/10/2011',
    Pearson: 1185,
    Sky: 729,
    Bloomsbury: 99.75,
    'Vitesse Media': 4.75,
    Totally: 47.3,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/11/2011',
    Pearson: 1148,
    Sky: 733,
    Bloomsbury: 95.13,
    'Vitesse Media': 4.75,
    Totally: 47.3,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/11/2011',
    Pearson: 1119,
    Sky: 738.5,
    Bloomsbury: 96.25,
    'Vitesse Media': 4.75,
    Totally: 47.3,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/11/2011',
    Pearson: 1102,
    Sky: 721,
    Bloomsbury: 97,
    'Vitesse Media': 4.75,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/11/2011',
    Pearson: 1069,
    Sky: 715.5,
    Bloomsbury: 95,
    'Vitesse Media': 4.75,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/12/2011',
    Pearson: 1135,
    Sky: 766,
    Bloomsbury: 94.75,
    'Vitesse Media': 4.75,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/12/2011',
    Pearson: 1133,
    Sky: 734.5,
    Bloomsbury: 97.25,
    'Vitesse Media': 4.75,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/12/2011',
    Pearson: 1134,
    Sky: 725,
    Bloomsbury: 98.38,
    'Vitesse Media': 4.75,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/12/2011',
    Pearson: 1191,
    Sky: 712.5,
    Bloomsbury: 99,
    'Vitesse Media': 4.75,
    Totally: 52.27,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/12/2011',
    Pearson: 1222,
    Sky: 734.5,
    Bloomsbury: 98.38,
    'Vitesse Media': 4.75,
    Totally: 52.27,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/01/2012',
    Pearson: 1212,
    Sky: 721,
    Bloomsbury: 95,
    'Vitesse Media': 4.75,
    Totally: 72.19,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/01/2012',
    Pearson: 1222,
    Sky: 697.5,
    Bloomsbury: 92.5,
    'Vitesse Media': 4.75,
    Totally: 72.19,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/01/2012',
    Pearson: 1230,
    Sky: 666.5,
    Bloomsbury: 103,
    'Vitesse Media': 4.25,
    Totally: 67.21,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/01/2012',
    Pearson: 1177,
    Sky: 671.5,
    Bloomsbury: 109.5,
    'Vitesse Media': 4.25,
    Totally: 74.68,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/02/2012',
    Pearson: 1196,
    Sky: 686,
    Bloomsbury: 112,
    'Vitesse Media': 4.25,
    Totally: 64.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/02/2012',
    Pearson: 1202,
    Sky: 701,
    Bloomsbury: 114,
    'Vitesse Media': 4.25,
    Totally: 151.84,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/02/2012',
    Pearson: 1221,
    Sky: 689.5,
    Bloomsbury: 124,
    'Vitesse Media': 4.25,
    Totally: 112.02,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/02/2012',
    Pearson: 1233,
    Sky: 687.5,
    Bloomsbury: 121.75,
    'Vitesse Media': 3.5,
    Totally: 107.04,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/03/2012',
    Pearson: 1211,
    Sky: 674.5,
    Bloomsbury: 121.25,
    'Vitesse Media': 3.75,
    Totally: 107.04,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/03/2012',
    Pearson: 1209,
    Sky: 694,
    Bloomsbury: 117.25,
    'Vitesse Media': 3.63,
    Totally: 109.53,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/03/2012',
    Pearson: 1210,
    Sky: 693.5,
    Bloomsbury: 115.25,
    'Vitesse Media': 3.75,
    Totally: 112.02,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/03/2012',
    Pearson: 1205,
    Sky: 693.5,
    Bloomsbury: 113.75,
    'Vitesse Media': 3.75,
    Totally: 97.08,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/03/2012',
    Pearson: 1170,
    Sky: 667,
    Bloomsbury: 109.5,
    'Vitesse Media': 3.75,
    Totally: 94.59,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/04/2012',
    Pearson: 1135,
    Sky: 635.5,
    Bloomsbury: 109.5,
    'Vitesse Media': 3.25,
    Totally: 92.1,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/04/2012',
    Pearson: 1134,
    Sky: 662,
    Bloomsbury: 109.87,
    'Vitesse Media': 3.25,
    Totally: 82.14,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/04/2012',
    Pearson: 1113,
    Sky: 665.5,
    Bloomsbury: 109.87,
    'Vitesse Media': 3.75,
    Totally: 87.12,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/04/2012',
    Pearson: 1150,
    Sky: 667.5,
    Bloomsbury: 106.75,
    'Vitesse Media': 3.75,
    Totally: 87.12,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/05/2012',
    Pearson: 1166,
    Sky: 704,
    Bloomsbury: 108,
    'Vitesse Media': 3.63,
    Totally: 84.63,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/05/2012',
    Pearson: 1163,
    Sky: 692,
    Bloomsbury: 109,
    'Vitesse Media': 3.63,
    Totally: 74.68,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/05/2012',
    Pearson: 1143,
    Sky: 689,
    Bloomsbury: 105,
    'Vitesse Media': 3.5,
    Totally: 72.19,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/05/2012',
    Pearson: 1137,
    Sky: 699.5,
    Bloomsbury: 115,
    'Vitesse Media': 3.5,
    Totally: 84.63,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '31/05/2012',
    Pearson: 1137,
    Sky: 688,
    Bloomsbury: 114,
    'Vitesse Media': 3.5,
    Totally: 92.1,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/06/2012',
    Pearson: 1148,
    Sky: 681,
    Bloomsbury: 112,
    'Vitesse Media': 3.5,
    Totally: 79.66,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/06/2012',
    Pearson: 1184,
    Sky: 671,
    Bloomsbury: 113,
    'Vitesse Media': 3.5,
    Totally: 74.68,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/06/2012',
    Pearson: 1210,
    Sky: 678.5,
    Bloomsbury: 113.25,
    'Vitesse Media': 3.5,
    Totally: 84.63,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/06/2012',
    Pearson: 1243,
    Sky: 680,
    Bloomsbury: 115,
    'Vitesse Media': 3.75,
    Totally: 79.66,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/07/2012',
    Pearson: 1250,
    Sky: 687.5,
    Bloomsbury: 115,
    'Vitesse Media': 3.75,
    Totally: 94.59,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/07/2012',
    Pearson: 1249,
    Sky: 688,
    Bloomsbury: 115,
    'Vitesse Media': 3.75,
    Totally: 116.99,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/07/2012',
    Pearson: 1290,
    Sky: 693.5,
    Bloomsbury: 128,
    'Vitesse Media': 3.75,
    Totally: 149.35,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/07/2012',
    Pearson: 1294,
    Sky: 706.5,
    Bloomsbury: 135,
    'Vitesse Media': 3.75,
    Totally: 109.53,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/08/2012',
    Pearson: 1223,
    Sky: 723.5,
    Bloomsbury: 129,
    'Vitesse Media': 3.75,
    Totally: 126.95,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/08/2012',
    Pearson: 1235,
    Sky: 752,
    Bloomsbury: 135,
    'Vitesse Media': 3.75,
    Totally: 126.95,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/08/2012',
    Pearson: 1212,
    Sky: 766,
    Bloomsbury: 140,
    'Vitesse Media': 3.75,
    Totally: 139.4,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/08/2012',
    Pearson: 1212,
    Sky: 760.5,
    Bloomsbury: 145.25,
    'Vitesse Media': 3.25,
    Totally: 136.91,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/08/2012',
    Pearson: 1201,
    Sky: 760,
    Bloomsbury: 136.25,
    'Vitesse Media': 3.25,
    Totally: 134.42,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/09/2012',
    Pearson: 1202,
    Sky: 769.5,
    Bloomsbury: 130.25,
    'Vitesse Media': 3.25,
    Totally: 121.97,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/09/2012',
    Pearson: 1179,
    Sky: 744,
    Bloomsbury: 134.5,
    'Vitesse Media': 3.25,
    Totally: 121.97,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/09/2012',
    Pearson: 1182,
    Sky: 734,
    Bloomsbury: 133,
    'Vitesse Media': 3.25,
    Totally: 159.31,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/09/2012',
    Pearson: 1217,
    Sky: 751,
    Bloomsbury: 134,
    'Vitesse Media': 3.25,
    Totally: 139.4,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/10/2012',
    Pearson: 1252,
    Sky: 761,
    Bloomsbury: 134.5,
    'Vitesse Media': 3.25,
    Totally: 129.44,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/10/2012',
    Pearson: 1248,
    Sky: 751,
    Bloomsbury: 131.25,
    'Vitesse Media': 3.13,
    Totally: 131.93,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/10/2012',
    Pearson: 1215,
    Sky: 733.5,
    Bloomsbury: 140,
    'Vitesse Media': 3.13,
    Totally: 149.35,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/10/2012',
    Pearson: 1212,
    Sky: 720,
    Bloomsbury: 131.75,
    'Vitesse Media': 3.13,
    Totally: 141.89,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/11/2012',
    Pearson: 1252,
    Sky: 759,
    Bloomsbury: 126.75,
    'Vitesse Media': 2.88,
    Totally: 129.44,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/11/2012',
    Pearson: 1230,
    Sky: 749.5,
    Bloomsbury: 123,
    'Vitesse Media': 2.88,
    Totally: 134.42,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/11/2012',
    Pearson: 1189,
    Sky: 746,
    Bloomsbury: 120,
    'Vitesse Media': 2.75,
    Totally: 129.44,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/11/2012',
    Pearson: 1175,
    Sky: 766,
    Bloomsbury: 120,
    'Vitesse Media': 2.75,
    Totally: 129.44,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/11/2012',
    Pearson: 1175,
    Sky: 771.5,
    Bloomsbury: 120,
    'Vitesse Media': 2.75,
    Totally: 146.86,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/12/2012',
    Pearson: 1190,
    Sky: 766.5,
    Bloomsbury: 118.75,
    'Vitesse Media': 2.63,
    Totally: 141.89,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/12/2012',
    Pearson: 1184,
    Sky: 761,
    Bloomsbury: 119.88,
    'Vitesse Media': 2.63,
    Totally: 126.95,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/12/2012',
    Pearson: 1200,
    Sky: 780,
    Bloomsbury: 119,
    'Vitesse Media': 2.63,
    Totally: 116.99,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/12/2012',
    Pearson: 1196,
    Sky: 780.5,
    Bloomsbury: 117,
    'Vitesse Media': 2.63,
    Totally: 109.53,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/01/2013',
    Pearson: 1211,
    Sky: 780.5,
    Bloomsbury: 118.25,
    'Vitesse Media': 2.63,
    Totally: 112.02,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/01/2013',
    Pearson: 1193,
    Sky: 773.5,
    Bloomsbury: 118,
    'Vitesse Media': 2.63,
    Totally: 161.8,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/01/2013',
    Pearson: 1232,
    Sky: 790.5,
    Bloomsbury: 121,
    'Vitesse Media': 2.63,
    Totally: 149.35,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/01/2013',
    Pearson: 1187,
    Sky: 794,
    Bloomsbury: 122.75,
    'Vitesse Media': 2.38,
    Totally: 144.37,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '31/01/2013',
    Pearson: 1194,
    Sky: 817.5,
    Bloomsbury: 115,
    'Vitesse Media': 2.13,
    Totally: 161.8,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/02/2013',
    Pearson: 1203,
    Sky: 818,
    Bloomsbury: 112,
    'Vitesse Media': 2.13,
    Totally: 156.82,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/02/2013',
    Pearson: 1207,
    Sky: 831,
    Bloomsbury: 110.13,
    'Vitesse Media': 2.25,
    Totally: 154.33,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/02/2013',
    Pearson: 1222,
    Sky: 824.5,
    Bloomsbury: 108,
    'Vitesse Media': 2.13,
    Totally: 139.4,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/02/2013',
    Pearson: 1155,
    Sky: 850.5,
    Bloomsbury: 104,
    'Vitesse Media': 1.88,
    Totally: 139.4,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/03/2013',
    Pearson: 1165,
    Sky: 866.5,
    Bloomsbury: 108,
    'Vitesse Media': 1.88,
    Totally: 121.97,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/03/2013',
    Pearson: 1204,
    Sky: 881,
    Bloomsbury: 106.75,
    'Vitesse Media': 2.13,
    Totally: 119.48,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/03/2013',
    Pearson: 1172,
    Sky: 886.5,
    Bloomsbury: 113,
    'Vitesse Media': 2.13,
    Totally: 136.91,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/03/2013',
    Pearson: 1184,
    Sky: 883,
    Bloomsbury: 115,
    'Vitesse Media': 1.65,
    Totally: 136.91,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/04/2013',
    Pearson: 1142,
    Sky: 878.5,
    Bloomsbury: 116.25,
    'Vitesse Media': 1.63,
    Totally: 136.91,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/04/2013',
    Pearson: 1155,
    Sky: 864,
    Bloomsbury: 115,
    'Vitesse Media': 1.63,
    Totally: 136.91,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/04/2013',
    Pearson: 1119,
    Sky: 836.5,
    Bloomsbury: 113.38,
    'Vitesse Media': 1.63,
    Totally: 136.91,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/04/2013',
    Pearson: 1149,
    Sky: 850,
    Bloomsbury: 114.5,
    'Vitesse Media': 2,
    Totally: 119.48,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/05/2013',
    Pearson: 1169,
    Sky: 851,
    Bloomsbury: 115,
    'Vitesse Media': 2,
    Totally: 144.37,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/05/2013',
    Pearson: 1186,
    Sky: 809,
    Bloomsbury: 117.5,
    'Vitesse Media': 2.13,
    Totally: 144.37,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/05/2013',
    Pearson: 1204,
    Sky: 795,
    Bloomsbury: 118,
    'Vitesse Media': 2.13,
    Totally: 166.78,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/05/2013',
    Pearson: 1217,
    Sky: 791.5,
    Bloomsbury: 121,
    'Vitesse Media': 2.13,
    Totally: 156.82,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/05/2013',
    Pearson: 1238,
    Sky: 782,
    Bloomsbury: 130,
    'Vitesse Media': 2.13,
    Totally: 164.29,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/06/2013',
    Pearson: 1177,
    Sky: 764,
    Bloomsbury: 126.25,
    'Vitesse Media': 2,
    Totally: 139.4,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/06/2013',
    Pearson: 1179,
    Sky: 792,
    Bloomsbury: 121,
    'Vitesse Media': 2.25,
    Totally: 129.44,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/06/2013',
    Pearson: 1162,
    Sky: 775,
    Bloomsbury: 125.75,
    'Vitesse Media': 2.25,
    Totally: 131.93,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/06/2013',
    Pearson: 1154,
    Sky: 790,
    Bloomsbury: 127.75,
    'Vitesse Media': 2.25,
    Totally: 121.97,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/07/2013',
    Pearson: 1206,
    Sky: 812.5,
    Bloomsbury: 127.5,
    'Vitesse Media': 2.13,
    Totally: 131.93,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/07/2013',
    Pearson: 1242,
    Sky: 826.5,
    Bloomsbury: 133.5,
    'Vitesse Media': 2.13,
    Totally: 129.44,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/07/2013',
    Pearson: 1264,
    Sky: 845,
    Bloomsbury: 137.5,
    'Vitesse Media': 2.13,
    Totally: 116.99,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/07/2013',
    Pearson: 1252,
    Sky: 850,
    Bloomsbury: 134,
    'Vitesse Media': 1.75,
    Totally: 109.53,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/08/2013',
    Pearson: 1344,
    Sky: 833,
    Bloomsbury: 139.5,
    'Vitesse Media': 1.75,
    Totally: 102.06,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/08/2013',
    Pearson: 1324,
    Sky: 841,
    Bloomsbury: 140.63,
    'Vitesse Media': 2,
    Totally: 89.61,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/08/2013',
    Pearson: 1313,
    Sky: 830,
    Bloomsbury: 140.5,
    'Vitesse Media': 1.63,
    Totally: 89.61,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/08/2013',
    Pearson: 1325,
    Sky: 827,
    Bloomsbury: 141,
    'Vitesse Media': 1.63,
    Totally: 82.14,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/08/2013',
    Pearson: 1280,
    Sky: 847,
    Bloomsbury: 137,
    'Vitesse Media': 1.63,
    Totally: 74.68,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/09/2013',
    Pearson: 1285,
    Sky: 851.5,
    Bloomsbury: 142.75,
    'Vitesse Media': 1.75,
    Totally: 79.66,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/09/2013',
    Pearson: 1275,
    Sky: 857.5,
    Bloomsbury: 144.25,
    'Vitesse Media': 2,
    Totally: 67.21,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/09/2013',
    Pearson: 1286,
    Sky: 861.5,
    Bloomsbury: 146.5,
    'Vitesse Media': 2,
    Totally: 64.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/09/2013',
    Pearson: 1253,
    Sky: 877.5,
    Bloomsbury: 147.25,
    'Vitesse Media': 2.25,
    Totally: 72.19,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/10/2013',
    Pearson: 1226,
    Sky: 881.5,
    Bloomsbury: 154.5,
    'Vitesse Media': 2.25,
    Totally: 72.19,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/10/2013',
    Pearson: 1297,
    Sky: 869.5,
    Bloomsbury: 165.75,
    'Vitesse Media': 2.13,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/10/2013',
    Pearson: 1317,
    Sky: 940,
    Bloomsbury: 168.5,
    'Vitesse Media': 2.13,
    Totally: 67.21,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/10/2013',
    Pearson: 1355,
    Sky: 950,
    Bloomsbury: 171.75,
    'Vitesse Media': 2.13,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '31/10/2013',
    Pearson: 1304,
    Sky: 937.5,
    Bloomsbury: 172.5,
    'Vitesse Media': 2.13,
    Totally: 64.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/11/2013',
    Pearson: 1293,
    Sky: 934.5,
    Bloomsbury: 174.75,
    'Vitesse Media': 2.13,
    Totally: 74.68,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/11/2013',
    Pearson: 1325,
    Sky: 818.5,
    Bloomsbury: 174.75,
    'Vitesse Media': 2.13,
    Totally: 77.17,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/11/2013',
    Pearson: 1342,
    Sky: 834,
    Bloomsbury: 168.5,
    'Vitesse Media': 2.13,
    Totally: 67.21,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/11/2013',
    Pearson: 1352,
    Sky: 819.5,
    Bloomsbury: 170.88,
    'Vitesse Media': 2.13,
    Totally: 69.7,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/12/2013',
    Pearson: 1287,
    Sky: 789.5,
    Bloomsbury: 171.75,
    'Vitesse Media': 2.13,
    Totally: 69.7,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/12/2013',
    Pearson: 1271,
    Sky: 788,
    Bloomsbury: 174.75,
    'Vitesse Media': 2.13,
    Totally: 67.21,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/12/2013',
    Pearson: 1300,
    Sky: 793,
    Bloomsbury: 177.75,
    'Vitesse Media': 2.13,
    Totally: 64.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/12/2013',
    Pearson: 1326,
    Sky: 833,
    Bloomsbury: 176.25,
    'Vitesse Media': 2.13,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/01/2014',
    Pearson: 1340,
    Sky: 843.5,
    Bloomsbury: 175.25,
    'Vitesse Media': 2.13,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/01/2014',
    Pearson: 1305,
    Sky: 824.5,
    Bloomsbury: 168,
    'Vitesse Media': 2.13,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/01/2014',
    Pearson: 1291,
    Sky: 845,
    Bloomsbury: 167.25,
    'Vitesse Media': 2.13,
    Totally: 52.27,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/01/2014',
    Pearson: 1191,
    Sky: 854.5,
    Bloomsbury: 168.5,
    'Vitesse Media': 2.13,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/01/2014',
    Pearson: 1115,
    Sky: 878,
    Bloomsbury: 173,
    'Vitesse Media': 2.75,
    Totally: 52.27,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/02/2014',
    Pearson: 1086,
    Sky: 893.5,
    Bloomsbury: 170,
    'Vitesse Media': 2.75,
    Totally: 52.27,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/02/2014',
    Pearson: 1115,
    Sky: 891.5,
    Bloomsbury: 177,
    'Vitesse Media': 2.75,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/02/2014',
    Pearson: 1110,
    Sky: 917,
    Bloomsbury: 178.75,
    'Vitesse Media': 2.75,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/02/2014',
    Pearson: 1076,
    Sky: 939.5,
    Bloomsbury: 174.38,
    'Vitesse Media': 3.63,
    Totally: 69.7,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/03/2014',
    Pearson: 1033,
    Sky: 932,
    Bloomsbury: 175.25,
    'Vitesse Media': 7,
    Totally: 72.19,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/03/2014',
    Pearson: 998.5,
    Sky: 914.5,
    Bloomsbury: 175.25,
    'Vitesse Media': 5.88,
    Totally: 64.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/03/2014',
    Pearson: 999.5,
    Sky: 942,
    Bloomsbury: 176,
    'Vitesse Media': 5,
    Totally: 69.7,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/03/2014',
    Pearson: 1026,
    Sky: 930,
    Bloomsbury: 174.75,
    'Vitesse Media': 4.63,
    Totally: 77.17,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/04/2014',
    Pearson: 1008,
    Sky: 892.5,
    Bloomsbury: 174,
    'Vitesse Media': 4.75,
    Totally: 69.7,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/04/2014',
    Pearson: 1041,
    Sky: 887.5,
    Bloomsbury: 169.25,
    'Vitesse Media': 4.75,
    Totally: 87.12,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/04/2014',
    Pearson: 1062,
    Sky: 890,
    Bloomsbury: 167.12,
    'Vitesse Media': 4.75,
    Totally: 74.68,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/04/2014',
    Pearson: 1050,
    Sky: 893.5,
    Bloomsbury: 169,
    'Vitesse Media': 4.75,
    Totally: 72.19,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/05/2014',
    Pearson: 1102,
    Sky: 900.5,
    Bloomsbury: 158.25,
    'Vitesse Media': 4.88,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/05/2014',
    Pearson: 1118,
    Sky: 889,
    Bloomsbury: 150.25,
    'Vitesse Media': 5,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/05/2014',
    Pearson: 1150,
    Sky: 850,
    Bloomsbury: 167,
    'Vitesse Media': 5,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/05/2014',
    Pearson: 1144,
    Sky: 872.5,
    Bloomsbury: 163.5,
    'Vitesse Media': 4.88,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/05/2014',
    Pearson: 1164,
    Sky: 885.5,
    Bloomsbury: 167,
    'Vitesse Media': 4.88,
    Totally: 69.7,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/06/2014',
    Pearson: 1148,
    Sky: 872,
    Bloomsbury: 167,
    'Vitesse Media': 4.88,
    Totally: 69.7,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/06/2014',
    Pearson: 1129,
    Sky: 876,
    Bloomsbury: 166.5,
    'Vitesse Media': 4.88,
    Totally: 64.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/06/2014',
    Pearson: 1123,
    Sky: 871.5,
    Bloomsbury: 188.5,
    'Vitesse Media': 4.88,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/06/2014',
    Pearson: 1144,
    Sky: 889.5,
    Bloomsbury: 178.75,
    'Vitesse Media': 4.75,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/07/2014',
    Pearson: 1143,
    Sky: 903,
    Bloomsbury: 177.75,
    'Vitesse Media': 4.75,
    Totally: 52.27,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/07/2014',
    Pearson: 1092,
    Sky: 869.5,
    Bloomsbury: 165.75,
    'Vitesse Media': 4.75,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/07/2014',
    Pearson: 1109,
    Sky: 897.5,
    Bloomsbury: 173.5,
    'Vitesse Media': 4.75,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/07/2014',
    Pearson: 1101,
    Sky: 925,
    Bloomsbury: 175,
    'Vitesse Media': 4.75,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '31/07/2014',
    Pearson: 1141,
    Sky: 879,
    Bloomsbury: 176.25,
    'Vitesse Media': 4.88,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/08/2014',
    Pearson: 1119,
    Sky: 850,
    Bloomsbury: 171.75,
    'Vitesse Media': 4.88,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/08/2014',
    Pearson: 1120,
    Sky: 854,
    Bloomsbury: 171,
    'Vitesse Media': 4.88,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/08/2014',
    Pearson: 1125,
    Sky: 861.5,
    Bloomsbury: 177.13,
    'Vitesse Media': 4.5,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/08/2014',
    Pearson: 1111,
    Sky: 871,
    Bloomsbury: 179.38,
    'Vitesse Media': 4.38,
    Totally: 54.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/09/2014',
    Pearson: 1109,
    Sky: 885,
    Bloomsbury: 173.5,
    'Vitesse Media': 3.75,
    Totally: 49.78,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/09/2014',
    Pearson: 1154,
    Sky: 868,
    Bloomsbury: 170.5,
    'Vitesse Media': 3.75,
    Totally: 49.78,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/09/2014',
    Pearson: 1218,
    Sky: 884,
    Bloomsbury: 164.25,
    'Vitesse Media': 3.75,
    Totally: 49.78,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/09/2014',
    Pearson: 1230,
    Sky: 872.5,
    Bloomsbury: 166,
    'Vitesse Media': 3.75,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/10/2014',
    Pearson: 1204,
    Sky: 875,
    Bloomsbury: 165.25,
    'Vitesse Media': 3.63,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/10/2014',
    Pearson: 1173,
    Sky: 892.5,
    Bloomsbury: 165,
    'Vitesse Media': 3.63,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/10/2014',
    Pearson: 1132,
    Sky: 850,
    Bloomsbury: 159.25,
    'Vitesse Media': 3.13,
    Totally: 32.36,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/10/2014',
    Pearson: 1169,
    Sky: 859.5,
    Bloomsbury: 154.25,
    'Vitesse Media': 3.13,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/10/2014',
    Pearson: 1160,
    Sky: 882,
    Bloomsbury: 157,
    'Vitesse Media': 3,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/11/2014',
    Pearson: 1186,
    Sky: 885,
    Bloomsbury: 162,
    'Vitesse Media': 3.13,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/11/2014',
    Pearson: 1201,
    Sky: 850.5,
    Bloomsbury: 161.25,
    'Vitesse Media': 3.13,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/11/2014',
    Pearson: 1220,
    Sky: 897,
    Bloomsbury: 163.88,
    'Vitesse Media': 3.13,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/11/2014',
    Pearson: 1224,
    Sky: 922.5,
    Bloomsbury: 162.75,
    'Vitesse Media': 3.75,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/12/2014',
    Pearson: 1212,
    Sky: 926,
    Bloomsbury: 161,
    'Vitesse Media': 3.13,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/12/2014',
    Pearson: 1170,
    Sky: 918,
    Bloomsbury: 160,
    'Vitesse Media': 3.13,
    Totally: 29.87,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/12/2014',
    Pearson: 1190,
    Sky: 900,
    Bloomsbury: 161,
    'Vitesse Media': 2.88,
    Totally: 29.87,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/12/2014',
    Pearson: 1179,
    Sky: 899,
    Bloomsbury: 162.5,
    'Vitesse Media': 2.88,
    Totally: 32.36,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/01/2015',
    Pearson: '',
    Sky: 899,
    Bloomsbury: 162.13,
    'Vitesse Media': 2.88,
    Totally: 32.36,
    highlight: 'begin',
    annotate: 'vertical annotation',
    type: 'vertical',
  },
  {
    date: '08/01/2015',
    Pearson: '',
    Sky: 887,
    Bloomsbury: 152.38,
    'Vitesse Media': 2.88,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/01/2015',
    Pearson: '',
    Sky: 919.5,
    Bloomsbury: 156.25,
    'Vitesse Media': 2.88,
    Totally: 24.89,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/01/2015',
    Pearson: '',
    Sky: 936.5,
    Bloomsbury: 152,
    'Vitesse Media': 2.88,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/01/2015',
    Pearson: '',
    Sky: 934,
    Bloomsbury: 149,
    'Vitesse Media': 2.88,
    Totally: 34.85,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/02/2015',
    Pearson: '',
    Sky: 960,
    Bloomsbury: 149.75,
    'Vitesse Media': 2.63,
    Totally: 32.36,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/02/2015',
    Pearson: '',
    Sky: 927.5,
    Bloomsbury: 151.25,
    'Vitesse Media': 2.63,
    Totally: 24.89,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/02/2015',
    Pearson: '',
    Sky: 956,
    Bloomsbury: 149.25,
    'Vitesse Media': 2.88,
    Totally: 27.38,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/02/2015',
    Pearson: '',
    Sky: 998.5,
    Bloomsbury: 148.75,
    'Vitesse Media': 2.75,
    Totally: 27.38,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/03/2015',
    Pearson: '',
    Sky: 1003,
    Bloomsbury: 153,
    'Vitesse Media': 2.5,
    Totally: 24.89,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/03/2015',
    Pearson: '',
    Sky: 991,
    Bloomsbury: 145.25,
    'Vitesse Media': 2.5,
    Totally: 24.89,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/03/2015',
    Pearson: '',
    Sky: 1027,
    Bloomsbury: 151.25,
    'Vitesse Media': 2.5,
    Totally: 22.4,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/03/2015',
    Pearson: '',
    Sky: 1006,
    Bloomsbury: 153.75,
    'Vitesse Media': 2.38,
    Totally: 17.42,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/04/2015',
    Pearson: '',
    Sky: 997,
    Bloomsbury: 150,
    'Vitesse Media': 2.38,
    Totally: 24.89,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/04/2015',
    Pearson: '',
    Sky: 1045,
    Bloomsbury: 159.75,
    'Vitesse Media': 2.38,
    Totally: 24.89,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/04/2015',
    Pearson: '',
    Sky: 1047,
    Bloomsbury: 160.25,
    'Vitesse Media': 2.38,
    Totally: 19.91,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/04/2015',
    Pearson: '',
    Sky: 1090,
    Bloomsbury: 161.75,
    'Vitesse Media': 2.25,
    Totally: 17.42,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/04/2015',
    Pearson: '',
    Sky: 1077,
    Bloomsbury: 167.5,
    'Vitesse Media': 2.25,
    Totally: 27.38,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/05/2015',
    Pearson: '',
    Sky: 1067,
    Bloomsbury: 173.5,
    'Vitesse Media': 2.25,
    Totally: 22.4,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/05/2015',
    Pearson: '',
    Sky: 1077,
    Bloomsbury: 178,
    'Vitesse Media': 2.13,
    Totally: 25.89,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/05/2015',
    Pearson: '',
    Sky: 1087,
    Bloomsbury: 177.25,
    'Vitesse Media': 2.13,
    Totally: 24.89,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/05/2015',
    Pearson: '',
    Sky: 1068,
    Bloomsbury: 184.5,
    'Vitesse Media': 3.13,
    Totally: 32.36,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/06/2015',
    Pearson: '',
    Sky: 1040,
    Bloomsbury: 179,
    'Vitesse Media': 3.25,
    Totally: 29.87,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/06/2015',
    Pearson: '',
    Sky: 1039,
    Bloomsbury: 177,
    'Vitesse Media': 3.13,
    Totally: 29.87,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/06/2015',
    Pearson: '',
    Sky: 1038,
    Bloomsbury: 176.5,
    'Vitesse Media': 3.13,
    Totally: 29.87,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/06/2015',
    Pearson: '',
    Sky: 1069,
    Bloomsbury: 168.25,
    'Vitesse Media': 3.13,
    Totally: 24.89,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/07/2015',
    Pearson: '',
    Sky: 1051,
    Bloomsbury: 161.25,
    'Vitesse Media': 2.75,
    Totally: 27.38,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/07/2015',
    Pearson: '',
    Sky: 1044,
    Bloomsbury: 158,
    'Vitesse Media': 2.75,
    Totally: 24.89,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/07/2015',
    Pearson: '',
    Sky: 1140,
    Bloomsbury: 162,
    'Vitesse Media': 2.25,
    Totally: 19.91,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/07/2015',
    Pearson: '',
    Sky: 1136,
    Bloomsbury: 163.25,
    'Vitesse Media': 2.13,
    Totally: 27.38,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/07/2015',
    Pearson: '',
    Sky: 1119,
    Bloomsbury: 164.13,
    'Vitesse Media': 1.5,
    Totally: 29.87,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/08/2015',
    Pearson: '',
    Sky: 1100,
    Bloomsbury: 164.5,
    'Vitesse Media': 1.63,
    Totally: 24.89,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/08/2015',
    Pearson: '',
    Sky: 1079,
    Bloomsbury: 169.63,
    'Vitesse Media': 1.88,
    Totally: 21.41,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/08/2015',
    Pearson: '',
    Sky: 1064,
    Bloomsbury: 173,
    'Vitesse Media': 1.75,
    Totally: 22.4,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/08/2015',
    Pearson: '',
    Sky: 1039,
    Bloomsbury: 170.75,
    'Vitesse Media': 1.75,
    Totally: 29.87,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/09/2015',
    Pearson: '',
    Sky: 1030,
    Bloomsbury: 169.25,
    'Vitesse Media': 1.75,
    Totally: 32.36,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/09/2015',
    Pearson: '',
    Sky: 1033,
    Bloomsbury: 165.5,
    'Vitesse Media': 1.63,
    Totally: 27.38,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/09/2015',
    Pearson: '',
    Sky: 1014,
    Bloomsbury: 168.75,
    'Vitesse Media': 1.88,
    Totally: 32.86,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/09/2015',
    Pearson: '',
    Sky: 1017,
    Bloomsbury: 169.5,
    'Vitesse Media': 1.88,
    Totally: 32.86,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/10/2015',
    Pearson: '',
    Sky: 1041,
    Bloomsbury: 169.25,
    'Vitesse Media': 1.88,
    Totally: 32.86,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/10/2015',
    Pearson: '',
    Sky: 1083,
    Bloomsbury: 169.5,
    'Vitesse Media': 1.63,
    Totally: 32.86,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/10/2015',
    Pearson: '',
    Sky: 1074,
    Bloomsbury: 169,
    'Vitesse Media': 1.75,
    Totally: 32.86,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/10/2015',
    Pearson: '',
    Sky: 1092,
    Bloomsbury: 168,
    'Vitesse Media': 1.75,
    Totally: 37.34,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/10/2015',
    Pearson: '',
    Sky: 1104,
    Bloomsbury: 167.25,
    'Vitesse Media': 1.25,
    Totally: 44.31,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/11/2015',
    Pearson: '',
    Sky: 1084,
    Bloomsbury: 167.75,
    'Vitesse Media': 1.25,
    Totally: 41.32,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/11/2015',
    Pearson: '',
    Sky: 1078,
    Bloomsbury: 163.5,
    'Vitesse Media': 1.25,
    Totally: 43.81,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/11/2015',
    Pearson: '',
    Sky: 1103,
    Bloomsbury: 160.25,
    'Vitesse Media': 1.13,
    Totally: 43.81,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/11/2015',
    Pearson: '',
    Sky: 1118,
    Bloomsbury: 159.5,
    'Vitesse Media': 1.13,
    Totally: 41.82,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/12/2015',
    Pearson: '',
    Sky: 1111,
    Bloomsbury: 160,
    'Vitesse Media': 1.13,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/12/2015',
    Pearson: '',
    Sky: 1076,
    Bloomsbury: 158.38,
    'Vitesse Media': 1.13,
    Totally: 39.83,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/12/2015',
    Pearson: '',
    Sky: 1102,
    Bloomsbury: 156.5,
    'Vitesse Media': 1.13,
    Totally: 41.82,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/12/2015',
    Pearson: '',
    Sky: 1099,
    Bloomsbury: 154,
    'Vitesse Media': 1.13,
    Totally: 41.82,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '31/12/2015',
    Pearson: '',
    Sky: 1112,
    Bloomsbury: 150.5,
    'Vitesse Media': 1.13,
    Totally: 41.82,
    highlight: 'end',
    annotate: '',
    type: '',
  },
  {
    date: '07/01/2016',
    Pearson: 701,
    Sky: 1084,
    Bloomsbury: 151.75,
    'Vitesse Media': 1.13,
    Totally: 72.69,
    highlight: '',
    annotate: 'annotation with curved line',
    type: 'curve',
  },
  {
    date: '14/01/2016',
    Pearson: 719.5,
    Sky: 1057,
    Bloomsbury: 147.25,
    'Vitesse Media': 1.25,
    Totally: 65.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/01/2016',
    Pearson: 772,
    Sky: 1012,
    Bloomsbury: 147,
    'Vitesse Media': 1.63,
    Totally: 68.2,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/01/2016',
    Pearson: 777,
    Sky: 1041,
    Bloomsbury: 145.5,
    'Vitesse Media': 1.38,
    Totally: 68.7,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/02/2016',
    Pearson: 780.5,
    Sky: 1066,
    Bloomsbury: 148.75,
    'Vitesse Media': 2.5,
    Totally: 68.2,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/02/2016',
    Pearson: 719.5,
    Sky: 976,
    Bloomsbury: 148.5,
    'Vitesse Media': 2.25,
    Totally: 63.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/02/2016',
    Pearson: 796.5,
    Sky: 1020,
    Bloomsbury: 153.5,
    'Vitesse Media': 1.88,
    Totally: 55.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/02/2016',
    Pearson: 801.5,
    Sky: 1026,
    Bloomsbury: 155.5,
    'Vitesse Media': 2.13,
    Totally: 56.75,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/03/2016',
    Pearson: 857,
    Sky: 1030,
    Bloomsbury: 153,
    'Vitesse Media': 2.13,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/03/2016',
    Pearson: 850,
    Sky: 984.5,
    Bloomsbury: 147.25,
    'Vitesse Media': 1.88,
    Totally: 69.7,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/03/2016',
    Pearson: 896.5,
    Sky: 1013,
    Bloomsbury: 150,
    'Vitesse Media': 1.63,
    Totally: 78.16,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/03/2016',
    Pearson: 883,
    Sky: 1016,
    Bloomsbury: 147.25,
    'Vitesse Media': 1.63,
    Totally: 70.69,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '31/03/2016',
    Pearson: 875,
    Sky: 1024,
    Bloomsbury: 147.25,
    'Vitesse Media': 1.63,
    Totally: 69.7,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/04/2016',
    Pearson: 825,
    Sky: 1019,
    Bloomsbury: 147,
    'Vitesse Media': 1.63,
    Totally: 71.19,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/04/2016',
    Pearson: 835.5,
    Sky: 1008,
    Bloomsbury: 148.13,
    'Vitesse Media': 1.75,
    Totally: 67.71,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/04/2016',
    Pearson: 834.5,
    Sky: 985,
    Bloomsbury: 156,
    'Vitesse Media': 1.63,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/04/2016',
    Pearson: 814.5,
    Sky: 944,
    Bloomsbury: 155.25,
    'Vitesse Media': 1.38,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/05/2016',
    Pearson: 805,
    Sky: 933.5,
    Bloomsbury: 152,
    'Vitesse Media': 2.38,
    Totally: 62.73,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/05/2016',
    Pearson: 806,
    Sky: 926,
    Bloomsbury: 150.88,
    'Vitesse Media': 2.25,
    Totally: 63.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/05/2016',
    Pearson: 802.5,
    Sky: 921,
    Bloomsbury: 151,
    'Vitesse Media': 2.38,
    Totally: 60.24,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/05/2016',
    Pearson: 825.5,
    Sky: 971.5,
    Bloomsbury: 163.5,
    'Vitesse Media': 2.75,
    Totally: 57.75,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/06/2016',
    Pearson: 849.5,
    Sky: 939,
    Bloomsbury: 164,
    'Vitesse Media': 2.75,
    Totally: 52.77,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/06/2016',
    Pearson: 818.5,
    Sky: 955,
    Bloomsbury: 161.38,
    'Vitesse Media': 2.88,
    Totally: 53.27,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/06/2016',
    Pearson: 807.5,
    Sky: 851.5,
    Bloomsbury: 151.5,
    'Vitesse Media': 3.13,
    Totally: 61.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/06/2016',
    Pearson: 888,
    Sky: 893.5,
    Bloomsbury: 161.5,
    'Vitesse Media': 2.88,
    Totally: 62.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/06/2016',
    Pearson: 967,
    Sky: 848,
    Bloomsbury: 160.25,
    'Vitesse Media': 2.75,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '07/07/2016',
    Pearson: 948,
    Sky: 848,
    Bloomsbury: 160,
    'Vitesse Media': 3.63,
    Totally: 62.73,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '14/07/2016',
    Pearson: 953.5,
    Sky: 872,
    Bloomsbury: 164.75,
    'Vitesse Media': 3.88,
    Totally: 62.73,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '21/07/2016',
    Pearson: 963,
    Sky: 867,
    Bloomsbury: 167.5,
    'Vitesse Media': 4.13,
    Totally: 67.21,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '28/07/2016',
    Pearson: 970,
    Sky: 904,
    Bloomsbury: 167.5,
    'Vitesse Media': 4.13,
    Totally: 65.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '04/08/2016',
    Pearson: 893.5,
    Sky: 909.5,
    Bloomsbury: 168.75,
    'Vitesse Media': 4.25,
    Totally: 61.73,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '11/08/2016',
    Pearson: 890,
    Sky: 884.5,
    Bloomsbury: 178.5,
    'Vitesse Media': 4.75,
    Totally: 61.73,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '18/08/2016',
    Pearson: 863,
    Sky: 855,
    Bloomsbury: 175.5,
    'Vitesse Media': 4.75,
    Totally: 58.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '25/08/2016',
    Pearson: 862,
    Sky: 873,
    Bloomsbury: 170,
    'Vitesse Media': 4.13,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/09/2016',
    Pearson: 858.5,
    Sky: 852.5,
    Bloomsbury: 163,
    'Vitesse Media': 4.13,
    Totally: 69.7,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/09/2016',
    Pearson: 797,
    Sky: 862,
    Bloomsbury: 163.5,
    'Vitesse Media': 3.88,
    Totally: 65.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/09/2016',
    Pearson: 780,
    Sky: 833,
    Bloomsbury: 159,
    'Vitesse Media': 3.63,
    Totally: 64.72,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/09/2016',
    Pearson: 774,
    Sky: 862,
    Bloomsbury: 155.75,
    'Vitesse Media': 3.38,
    Totally: 67.21,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/09/2016',
    Pearson: 749.5,
    Sky: 910.5,
    Bloomsbury: 156.25,
    'Vitesse Media': 3.38,
    Totally: 64.22,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/10/2016',
    Pearson: 794,
    Sky: 901,
    Bloomsbury: 153,
    'Vitesse Media': 4.13,
    Totally: 55.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '13/10/2016',
    Pearson: 814,
    Sky: 848,
    Bloomsbury: 150,
    'Vitesse Media': 3.75,
    Totally: 61.23,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '20/10/2016',
    Pearson: 765.5,
    Sky: 825,
    Bloomsbury: 149,
    'Vitesse Media': 3.75,
    Totally: 58.75,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '27/10/2016',
    Pearson: 743,
    Sky: 820.5,
    Bloomsbury: 149.75,
    'Vitesse Media': 3.88,
    Totally: 56.75,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '03/11/2016',
    Pearson: 736,
    Sky: 811,
    Bloomsbury: 152,
    'Vitesse Media': 3.88,
    Totally: 59.24,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '10/11/2016',
    Pearson: 749.5,
    Sky: 790,
    Bloomsbury: 153.5,
    'Vitesse Media': 3.75,
    Totally: 57.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '17/11/2016',
    Pearson: 760.5,
    Sky: 784.5,
    Bloomsbury: 156.5,
    'Vitesse Media': 3.13,
    Totally: 56.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '24/11/2016',
    Pearson: 783.5,
    Sky: 757.5,
    Bloomsbury: 157.25,
    'Vitesse Media': 3.13,
    Totally: 57.75,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '01/12/2016',
    Pearson: 782.5,
    Sky: 773.5,
    Bloomsbury: 163.75,
    'Vitesse Media': 3.13,
    Totally: 59.74,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '08/12/2016',
    Pearson: 780.5,
    Sky: 789.5,
    Bloomsbury: 160.75,
    'Vitesse Media': 3.25,
    Totally: 56.75,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '15/12/2016',
    Pearson: 815,
    Sky: 988,
    Bloomsbury: 165.75,
    'Vitesse Media': 3.25,
    Totally: 57.75,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '22/12/2016',
    Pearson: 819,
    Sky: 988.5,
    Bloomsbury: 167,
    'Vitesse Media': 3.38,
    Totally: 53.77,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '29/12/2016',
    Pearson: 818.5,
    Sky: 989,
    Bloomsbury: 168.25,
    'Vitesse Media': 2.88,
    Totally: 53.77,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '05/01/2017',
    Pearson: 812.5,
    Sky: 990,
    Bloomsbury: 169.25,
    'Vitesse Media': 2.88,
    Totally: 54.27,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '12/01/2017',
    Pearson: 813,
    Sky: 990.5,
    Bloomsbury: 166,
    'Vitesse Media': 2.63,
    Totally: 55.76,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '19/01/2017',
    Pearson: 588.5,
    Sky: 999,
    Bloomsbury: 163,
    'Vitesse Media': 2.75,
    Totally: 56.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '26/01/2017',
    Pearson: 622,
    Sky: 1004,
    Bloomsbury: 169,
    'Vitesse Media': 2.88,
    Totally: 56.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/02/2017',
    Pearson: 643,
    Sky: 1005,
    Bloomsbury: 169,
    'Vitesse Media': 2.88,
    Totally: 56.26,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/02/2017',
    Pearson: 660,
    Sky: 1000,
    Bloomsbury: 170.63,
    'Vitesse Media': 2.88,
    Totally: 58.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/02/2017',
    Pearson: 660,
    Sky: 998,
    Bloomsbury: 170,
    'Vitesse Media': 2.88,
    Totally: 58.25,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/02/2017',
    Pearson: 646,
    Sky: 1000,
    Bloomsbury: 170.75,
    'Vitesse Media': 2.88,
    Totally: 57.75,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '02/03/2017',
    Pearson: 687.5,
    Sky: 998.5,
    Bloomsbury: 172.5,
    'Vitesse Media': 2.88,
    Totally: 58.5,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '09/03/2017',
    Pearson: 678,
    Sky: 996.5,
    Bloomsbury: 171.88,
    'Vitesse Media': 2.88,
    Totally: 59,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '16/03/2017',
    Pearson: 659,
    Sky: 991,
    Bloomsbury: 172.5,
    'Vitesse Media': 2.88,
    Totally: 59,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '23/03/2017',
    Pearson: 635,
    Sky: 982.5,
    Bloomsbury: 174,
    'Vitesse Media': 2.88,
    Totally: 59,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '30/03/2017',
    Pearson: 668,
    Sky: 977,
    Bloomsbury: 177.75,
    'Vitesse Media': 2.88,
    Totally: 59.5,
    highlight: '',
    annotate: '',
    type: '',
  },
  {
    date: '06/04/2017',
    Pearson: 637,
    Sky: 962,
    Bloomsbury: 172.5,
    'Vitesse Media': 2.88,
    Totally: 58.5,
    highlight: '',
    annotate: '',
    type: '',
  },
];

export default dateContinuousSampleData;
