// @financial-times/vvc/src/sample-data/date-continuous-log.js
const dateContinuousLogSampleData = [
  {
    date: 0,
    Algeria: 3.5714285714285716,
    Argentina: 3.2857142857142856,
    Australia: 3.2857142857142856,
    Austria: 3.142857142857143,
    Bangladesh: 3,
    Belarus: 3.2857142857142856,
    Belgium: 4.857142857142857,
    'Bosnia and Herzegovina': 3.142857142857143,
    Brazil: 3.5714285714285716,
    Canada: 3.7142857142857144,
    Chile: 3,
    China: 3.4285714285714284,
    Colombia: 3.7142857142857144,
    Czechia: 3.2857142857142856,
    Denmark: 3.2857142857142856,
    'Dominican Republic': 3.5714285714285716,
    Ecuador: 3.5714285714285716,
    Egypt: 3.142857142857143,
    Finland: 3.2857142857142856,
    France: 3.857142857142857,
    Germany: 5.428571428571429,
    Greece: 3,
    Hungary: 3,
    India: 3,
    Indonesia: 3,
    Iran: 3.4285714285714284,
    Iraq: 3.4285714285714284,
    Ireland: 4.714285714285714,
    Israel: 3.7142857142857144,
    Italy: 3,
    Japan: 3.4285714285714284,
    Luxembourg: 3,
    Malaysia: 3,
    Mexico: 3.4285714285714284,
    Moldova: 3.142857142857143,
    Morocco: 3.142857142857143,
    Netherlands: 3,
    'North Macedonia': 3,
    Norway: 4,
    Pakistan: 3.2857142857142856,
    Panama: 3,
    Peru: 3.2857142857142856,
    Philippines: 3,
    Poland: 3.2857142857142856,
    Portugal: 3.2857142857142856,
    'Puerto Rico': 3,
    Romania: 3.4285714285714284,
    Russia: 3.4285714285714284,
    'Saudi Arabia': 3.5714285714285716,
    Serbia: 3.142857142857143,
    Slovenia: 3.2857142857142856,
    'South Africa': 4.285714285714286,
    'South Korea': 3,
    Spain: 4,
    Sweden: 4,
    Switzerland: 4.142857142857143,
    Turkey: 3,
    Ukraine: 3.2857142857142856,
    'United Arab Emirates': 3,
    'United Kingdom': 4.571428571428571,
    'United States of America': 4,
  },
  {
    date: 1,
    Algeria: 5.285714285714286,
    Argentina: 3.142857142857143,
    Australia: 3.5714285714285716,
    Austria: 3.857142857142857,
    Bangladesh: 3.142857142857143,
    Belarus: 3.4285714285714284,
    Belgium: 9,
    'Bosnia and Herzegovina': 2.7142857142857144,
    Brazil: 4.857142857142857,
    Canada: 4.142857142857143,
    Chile: 3.857142857142857,
    China: 5.571428571428571,
    Colombia: 3.5714285714285716,
    Czechia: 4,
    Denmark: 4,
    'Dominican Republic': 5.142857142857143,
    Ecuador: 3.7142857142857144,
    Egypt: 3.7142857142857144,
    Finland: 3.2857142857142856,
    France: 4.142857142857143,
    Germany: 5.714285714285714,
    Greece: 2.857142857142857,
    Hungary: 3,
    India: 3.142857142857143,
    Indonesia: 4.857142857142857,
    Iran: 4.285714285714286,
    Iraq: 3.7142857142857144,
    Ireland: 6,
    Israel: 4.142857142857143,
    Italy: 3.857142857142857,
    Japan: 3.4285714285714284,
    Luxembourg: 3,
    Malaysia: 3.4285714285714284,
    Mexico: 3.4285714285714284,
    Moldova: 3.2857142857142856,
    Morocco: 4.142857142857143,
    Netherlands: 5.571428571428571,
    'North Macedonia': 2.857142857142857,
    Norway: 4,
    Pakistan: 3.142857142857143,
    Panama: 2.857142857142857,
    Peru: 5.571428571428571,
    Philippines: 3,
    Poland: 3.2857142857142856,
    Portugal: 4.571428571428571,
    'Puerto Rico': 3.4285714285714284,
    Romania: 4.142857142857143,
    Russia: 4,
    'Saudi Arabia': 3.7142857142857144,
    Serbia: 3,
    Slovenia: 3.5714285714285716,
    'South Korea': 3.2857142857142856,
    Spain: 5,
    Sweden: 4.571428571428571,
    Switzerland: 5.142857142857143,
    Turkey: 4.285714285714286,
    Ukraine: 3.857142857142857,
    'United Arab Emirates': 2.7142857142857144,
    'United Kingdom': 7.142857142857143,
    'United States of America': 4.714285714285714,
  },
  {
    date: 2,
    Algeria: 5.428571428571429,
    Argentina: 2.857142857142857,
    Australia: 4.142857142857143,
    Austria: 4.285714285714286,
    Bangladesh: 3.5714285714285716,
    Belgium: 10.142857142857142,
    'Bosnia and Herzegovina': 2.7142857142857144,
    Brazil: 6.428571428571429,
    Canada: 5.857142857142857,
    Chile: 4.142857142857143,
    China: 7.571428571428571,
    Colombia: 4.571428571428571,
    Czechia: 4.714285714285714,
    Denmark: 4.285714285714286,
    'Dominican Republic': 5.571428571428571,
    Ecuador: 4.428571428571429,
    Egypt: 3.7142857142857144,
    Finland: 4,
    France: 6.285714285714286,
    Germany: 8.428571428571429,
    Greece: 2.7142857142857144,
    Hungary: 3.2857142857142856,
    India: 3.2857142857142856,
    Indonesia: 6.142857142857143,
    Iran: 5.428571428571429,
    Iraq: 3.5714285714285716,
    Ireland: 6.857142857142857,
    Israel: 4.857142857142857,
    Italy: 4.714285714285714,
    Japan: 3.4285714285714284,
    Luxembourg: 2.2857142857142856,
    Malaysia: 3.2857142857142856,
    Mexico: 4.571428571428571,
    Moldova: 3,
    Morocco: 4.428571428571429,
    Netherlands: 7.571428571428571,
    'North Macedonia': 2.4285714285714284,
    Norway: 4.285714285714286,
    Pakistan: 3.142857142857143,
    Panama: 3.4285714285714284,
    Peru: 6.571428571428571,
    Philippines: 3,
    Poland: 4.142857142857143,
    Portugal: 5.857142857142857,
    'Puerto Rico': 3.4285714285714284,
    Romania: 5.428571428571429,
    Russia: 4.285714285714286,
    'Saudi Arabia': 4.285714285714286,
    Serbia: 4,
    Slovenia: 3.857142857142857,
    'South Korea': 4.142857142857143,
    Spain: 6.571428571428571,
    Sweden: 8,
    Switzerland: 6.428571428571429,
    Turkey: 5.285714285714286,
    Ukraine: 4.571428571428571,
    'United Kingdom': 7.714285714285714,
    'United States of America': 5.714285714285714,
  },
  {
    date: 3,
    Algeria: 11.285714285714286,
    Argentina: 3.4285714285714284,
    Australia: 4.142857142857143,
    Austria: 6.571428571428571,
    Bangladesh: 3.857142857142857,
    Belgium: 11.857142857142858,
    'Bosnia and Herzegovina': 2.2857142857142856,
    Brazil: 7.571428571428571,
    Canada: 6,
    Chile: 4.428571428571429,
    China: 11.142857142857142,
    Colombia: 4.857142857142857,
    Czechia: 5,
    Denmark: 5,
    'Dominican Republic': 6.428571428571429,
    Ecuador: 4.857142857142857,
    Egypt: 3.142857142857143,
    Finland: 3.4285714285714284,
    France: 7.714285714285714,
    Germany: 11.714285714285714,
    Greece: 3.2857142857142856,
    Hungary: 4.428571428571429,
    India: 3.7142857142857144,
    Indonesia: 6.285714285714286,
    Iran: 6.571428571428571,
    Iraq: 3.142857142857143,
    Ireland: 9.142857142857142,
    Israel: 4.857142857142857,
    Italy: 6.571428571428571,
    Japan: 3.142857142857143,
    Luxembourg: 1.8571428571428572,
    Malaysia: 3.4285714285714284,
    Mexico: 6,
    Moldova: 2.5714285714285716,
    Morocco: 4.714285714285714,
    Netherlands: 10.142857142857142,
    'North Macedonia': 2.4285714285714284,
    Norway: 5.142857142857143,
    Pakistan: 3.4285714285714284,
    Panama: 3.7142857142857144,
    Peru: 7.142857142857143,
    Philippines: 3.857142857142857,
    Poland: 5.857142857142857,
    Portugal: 8.142857142857142,
    'Puerto Rico': 3.4285714285714284,
    Romania: 5.285714285714286,
    Russia: 5.428571428571429,
    'Saudi Arabia': 4.428571428571429,
    Serbia: 4.714285714285714,
    Slovenia: 3.5714285714285716,
    'South Korea': 4,
    Spain: 11.571428571428571,
    Sweden: 10.857142857142858,
    Switzerland: 6.714285714285714,
    Turkey: 6.285714285714286,
    Ukraine: 4.571428571428571,
    'United Kingdom': 13.857142857142858,
    'United States of America': 6.857142857142857,
  },
  {
    date: 4,
    Algeria: 14.428571428571429,
    Argentina: 3.7142857142857144,
    Australia: 3.4285714285714284,
    Austria: 8.857142857142858,
    Bangladesh: 4.142857142857143,
    Belgium: 16.714285714285715,
    'Bosnia and Herzegovina': 2,
    Brazil: 10.142857142857142,
    Canada: 5.857142857142857,
    Chile: 4.571428571428571,
    China: 14.285714285714286,
    Colombia: 5.428571428571429,
    Czechia: 6.285714285714286,
    Denmark: 6.142857142857143,
    'Dominican Republic': 6.714285714285714,
    Ecuador: 5.857142857142857,
    Egypt: 3.5714285714285716,
    Finland: 4.142857142857143,
    France: 10,
    Germany: 16.142857142857142,
    Greece: 3.7142857142857144,
    Hungary: 5.428571428571429,
    India: 5.285714285714286,
    Indonesia: 7.142857142857143,
    Iran: 7.714285714285714,
    Iraq: 3.2857142857142856,
    Ireland: 10.857142857142858,
    Israel: 5.714285714285714,
    Italy: 9.857142857142858,
    Japan: 2.7142857142857144,
    Luxembourg: 2.142857142857143,
    Malaysia: 3.2857142857142856,
    Mexico: 6.857142857142857,
    Moldova: 2.4285714285714284,
    Morocco: 4.857142857142857,
    Netherlands: 14.428571428571429,
    'North Macedonia': 2.4285714285714284,
    Norway: 4.714285714285714,
    Pakistan: 5.285714285714286,
    Panama: 4,
    Peru: 8.142857142857142,
    Philippines: 5.142857142857143,
    Poland: 7.857142857142857,
    Portugal: 10,
    'Puerto Rico': 3.4285714285714284,
    Romania: 8.285714285714286,
    Russia: 5.285714285714286,
    'Saudi Arabia': 3.5714285714285716,
    Serbia: 4.857142857142857,
    Slovenia: 4,
    'South Korea': 4.714285714285714,
    Spain: 16.571428571428573,
    Sweden: 11.714285714285714,
    Switzerland: 7.428571428571429,
    Turkey: 8.285714285714286,
    Ukraine: 5.285714285714286,
    'United Kingdom': 19.142857142857142,
    'United States of America': 8.428571428571429,
  },
  {
    date: 5,
    Algeria: 17.285714285714285,
    Argentina: 4.142857142857143,
    Australia: 3.2857142857142856,
    Austria: 8.571428571428571,
    Bangladesh: 4.285714285714286,
    Belgium: 23.428571428571427,
    'Bosnia and Herzegovina': 1.4285714285714286,
    Brazil: 11.571428571428571,
    Canada: 9.285714285714286,
    Chile: 5.571428571428571,
    China: 16.428571428571427,
    Colombia: 7.142857142857143,
    Czechia: 6.857142857142857,
    Denmark: 7.428571428571429,
    'Dominican Republic': 7.142857142857143,
    Ecuador: 6.285714285714286,
    Egypt: 4.428571428571429,
    Finland: 4.571428571428571,
    France: 11.571428571428571,
    Germany: 19.428571428571427,
    Greece: 4.142857142857143,
    Hungary: 6.428571428571429,
    India: 5.571428571428571,
    Indonesia: 7.571428571428571,
    Iran: 8.857142857142858,
    Iraq: 3.2857142857142856,
    Ireland: 11.285714285714286,
    Israel: 6.428571428571429,
    Italy: 13.571428571428571,
    Japan: 4,
    Luxembourg: 2.7142857142857144,
    Malaysia: 3.142857142857143,
    Mexico: 9,
    Moldova: 2.4285714285714284,
    Morocco: 3,
    Netherlands: 17.714285714285715,
    'North Macedonia': 2.5714285714285716,
    Norway: 5.857142857142857,
    Pakistan: 5.142857142857143,
    Panama: 3.857142857142857,
    Peru: 9.285714285714286,
    Philippines: 6.142857142857143,
    Poland: 8.714285714285714,
    Portugal: 12.571428571428571,
    'Puerto Rico': 3.142857142857143,
    Romania: 10.285714285714286,
    Russia: 5.285714285714286,
    'Saudi Arabia': 3.2857142857142856,
    Serbia: 5.428571428571429,
    Slovenia: 3.5714285714285716,
    'South Korea': 4.142857142857143,
    Spain: 18.714285714285715,
    Sweden: 12.714285714285714,
    Switzerland: 9.571428571428571,
    Turkey: 10.142857142857142,
    Ukraine: 6.571428571428571,
    'United Kingdom': 23.857142857142858,
    'United States of America': 11.428571428571429,
  },
  {
    date: 6,
    Algeria: 19.714285714285715,
    Argentina: 5.142857142857143,
    Australia: 3.5714285714285716,
    Austria: 10,
    Bangladesh: 5.571428571428571,
    Belgium: 28.428571428571427,
    'Bosnia and Herzegovina': 1,
    Brazil: 13.714285714285714,
    Canada: 9.857142857142858,
    Chile: 6.142857142857143,
    China: 21.857142857142858,
    Colombia: 7.857142857142857,
    Czechia: 7.285714285714286,
    Denmark: 8.428571428571429,
    'Dominican Republic': 6.857142857142857,
    Ecuador: 6.285714285714286,
    Egypt: 4,
    Finland: 4.285714285714286,
    France: 15.428571428571429,
    Germany: 26.428571428571427,
    Greece: 4,
    Hungary: 7.285714285714286,
    India: 7,
    Indonesia: 8.428571428571429,
    Iran: 10.428571428571429,
    Iraq: 3,
    Ireland: 14,
    Israel: 7.142857142857143,
    Italy: 18.714285714285715,
    Japan: 4.142857142857143,
    Luxembourg: 3,
    Malaysia: 3.7142857142857144,
    Mexico: 10.571428571428571,
    Moldova: 2.5714285714285716,
    Morocco: 5.142857142857143,
    Netherlands: 22.714285714285715,
    'North Macedonia': 2.142857142857143,
    Norway: 6.857142857142857,
    Pakistan: 4.571428571428571,
    Panama: 4.142857142857143,
    Peru: 9.714285714285714,
    Philippines: 6.571428571428571,
    Poland: 10.285714285714286,
    Portugal: 15,
    'Puerto Rico': 3.857142857142857,
    Romania: 11,
    Russia: 5.857142857142857,
    'Saudi Arabia': 3.7142857142857144,
    Serbia: 6.428571428571429,
    Slovenia: 3.5714285714285716,
    'South Korea': 3.7142857142857144,
    Spain: 40.42857142857143,
    Sweden: 17.285714285714285,
    Switzerland: 11.714285714285714,
    Turkey: 11.857142857142858,
    Ukraine: 6.428571428571429,
    'United Kingdom': 30.285714285714285,
    'United States of America': 17.142857142857142,
  },
  {
    date: 7,
    Algeria: 21.285714285714285,
    Argentina: 4.857142857142857,
    Australia: 2.7142857142857144,
    Austria: 11.857142857142858,
    Belgium: 36,
    'Bosnia and Herzegovina': 0.7142857142857143,
    Brazil: 15.857142857142858,
    Canada: 10.571428571428571,
    Chile: 6.571428571428571,
    China: 26.714285714285715,
    Colombia: 9.714285714285714,
    Czechia: 7.714285714285714,
    Denmark: 7.571428571428571,
    'Dominican Republic': 7,
    Ecuador: 7.428571428571429,
    Egypt: 5.142857142857143,
    Finland: 4.571428571428571,
    France: 16.857142857142858,
    Germany: 30,
    Greece: 3.857142857142857,
    Hungary: 7.571428571428571,
    India: 7.428571428571429,
    Indonesia: 8.857142857142858,
    Iran: 11.571428571428571,
    Iraq: 2.5714285714285716,
    Ireland: 14.428571428571429,
    Israel: 7.428571428571429,
    Italy: 25.142857142857142,
    Japan: 4.142857142857143,
    Luxembourg: 2.4285714285714284,
    Malaysia: 3.857142857142857,
    Mexico: 13.857142857142858,
    Moldova: 2.7142857142857144,
    Morocco: 6.285714285714286,
    Netherlands: 27,
    'North Macedonia': 2,
    Norway: 6.571428571428571,
    Pakistan: 4.571428571428571,
    Panama: 4.285714285714286,
    Peru: 11,
    Philippines: 7.857142857142857,
    Poland: 10.857142857142858,
    Portugal: 16.714285714285715,
    'Puerto Rico': 3.2857142857142856,
    Romania: 15.571428571428571,
    Russia: 5.571428571428571,
    'Saudi Arabia': 3.2857142857142856,
    Serbia: 5.428571428571429,
    Slovenia: 2.857142857142857,
    'South Korea': 4,
    Spain: 40.142857142857146,
    Sweden: 20.571428571428573,
    Switzerland: 18.285714285714285,
    Turkey: 12.428571428571429,
    Ukraine: 7.285714285714286,
    'United Kingdom': 35.142857142857146,
    'United States of America': 15.714285714285714,
  },
  {
    date: 8,
    Algeria: 21,
    Argentina: 6.428571428571429,
    Australia: 2.2857142857142856,
    Austria: 14,
    Belgium: 40.857142857142854,
    'Bosnia and Herzegovina': 0.7142857142857143,
    Brazil: 17.857142857142858,
    Canada: 14.142857142857142,
    Chile: 6.571428571428571,
    China: 31.142857142857142,
    Colombia: 10.571428571428571,
    Czechia: 8.142857142857142,
    Denmark: 8.285714285714286,
    'Dominican Republic': 6.142857142857143,
    Ecuador: 13,
    Egypt: 5,
    Finland: 4.714285714285714,
    France: 20.285714285714285,
    Germany: 40,
    Greece: 4.428571428571429,
    Hungary: 9.285714285714286,
    India: 11.428571428571429,
    Indonesia: 9.142857142857142,
    Iran: 12.857142857142858,
    Iraq: 2,
    Ireland: 16,
    Israel: 8,
    Italy: 29.142857142857142,
    Japan: 5.571428571428571,
    Luxembourg: 3.142857142857143,
    Malaysia: 3.857142857142857,
    Mexico: 16,
    Moldova: 3.5714285714285716,
    Morocco: 6.714285714285714,
    Netherlands: 33.285714285714285,
    Norway: 6.857142857142857,
    Pakistan: 5,
    Panama: 4.142857142857143,
    Peru: 10.571428571428571,
    Philippines: 7.571428571428571,
    Poland: 13.714285714285714,
    Portugal: 18.142857142857142,
    Romania: 16,
    Russia: 6.571428571428571,
    'Saudi Arabia': 3.5714285714285716,
    Serbia: 6,
    Slovenia: 3,
    'South Korea': 4.428571428571429,
    Spain: 65.14285714285714,
    Sweden: 28.142857142857142,
    Switzerland: 22,
    Turkey: 14.428571428571429,
    Ukraine: 7.857142857142857,
    'United Kingdom': 40,
    'United States of America': 30.428571428571427,
  },
  {
    date: 9,
    Algeria: 24.571428571428573,
    Argentina: 6.428571428571429,
    Australia: 1.8571428571428572,
    Austria: 16,
    Belgium: 50.857142857142854,
    Brazil: 22.142857142857142,
    Canada: 19.142857142857142,
    Chile: 6.428571428571429,
    China: 35.42857142857143,
    Colombia: 9.428571428571429,
    Czechia: 8.571428571428571,
    Denmark: 10,
    'Dominican Republic': 6.285714285714286,
    Ecuador: 12.285714285714286,
    Egypt: 5.428571428571429,
    France: 28,
    Germany: 46,
    Greece: 5.142857142857143,
    Hungary: 10.142857142857142,
    India: 11.714285714285714,
    Indonesia: 9.428571428571429,
    Iran: 14.571428571428571,
    Iraq: 2,
    Ireland: 17.142857142857142,
    Israel: 7.857142857142857,
    Italy: 47.285714285714285,
    Japan: 7.857142857142857,
    Luxembourg: 3.2857142857142856,
    Malaysia: 4.285714285714286,
    Mexico: 19.571428571428573,
    Morocco: 7.714285714285714,
    Netherlands: 42.57142857142857,
    Norway: 6.857142857142857,
    Pakistan: 5.714285714285714,
    Panama: 4.142857142857143,
    Peru: 11.857142857142858,
    Philippines: 8.285714285714286,
    Poland: 16.571428571428573,
    Portugal: 20.571428571428573,
    Romania: 15.428571428571429,
    Russia: 8.571428571428571,
    'Saudi Arabia': 3.857142857142857,
    Serbia: 5,
    Slovenia: 2.5714285714285716,
    'South Korea': 3.5714285714285716,
    Spain: 78.71428571428571,
    Sweden: 30.857142857142858,
    Switzerland: 25.571428571428573,
    Turkey: 18.714285714285715,
    Ukraine: 7.571428571428571,
    'United Kingdom': 51.714285714285715,
    'United States of America': 40.42857142857143,
  },
  {
    date: 10,
    Algeria: 21.571428571428573,
    Argentina: 6.571428571428571,
    Australia: 1.5714285714285714,
    Austria: 15.142857142857142,
    Belgium: 60.714285714285715,
    Brazil: 26.285714285714285,
    Canada: 24.428571428571427,
    Chile: 7,
    China: 40,
    Colombia: 11,
    Czechia: 9.714285714285714,
    Denmark: 11.714285714285714,
    'Dominican Republic': 6.714285714285714,
    Ecuador: 14.857142857142858,
    Egypt: 6.285714285714286,
    France: 44.42857142857143,
    Germany: 51.57142857142857,
    Greece: 5,
    Hungary: 10.714285714285714,
    India: 16.285714285714285,
    Indonesia: 10.428571428571429,
    Iran: 20,
    Iraq: 2.7142857142857144,
    Ireland: 19.857142857142858,
    Israel: 7.714285714285714,
    Italy: 58.857142857142854,
    Japan: 9,
    Luxembourg: 4.428571428571429,
    Malaysia: 3.857142857142857,
    Mexico: 20.571428571428573,
    Morocco: 7.714285714285714,
    Netherlands: 51.142857142857146,
    Norway: 6.428571428571429,
    Pakistan: 7.285714285714286,
    Panama: 4.428571428571429,
    Peru: 15.428571428571429,
    Philippines: 8.857142857142858,
    Poland: 16.714285714285715,
    Portugal: 21.285714285714285,
    Romania: 16.142857142857142,
    Russia: 9,
    'Saudi Arabia': 4.571428571428571,
    Serbia: 4.571428571428571,
    'South Korea': 4,
    Spain: 97.57142857142857,
    Sweden: 34.42857142857143,
    Switzerland: 28.142857142857142,
    Turkey: 24.285714285714285,
    Ukraine: 9.142857142857142,
    'United Kingdom': 51.42857142857143,
    'United States of America': 57.42857142857143,
  },
  {
    date: 11,
    Algeria: 20.714285714285715,
    Argentina: 7,
    Austria: 14.285714285714286,
    Belgium: 83.28571428571429,
    Brazil: 31.714285714285715,
    Canada: 31.285714285714285,
    Chile: 6.571428571428571,
    China: 45.714285714285715,
    Colombia: 10.857142857142858,
    Czechia: 9.428571428571429,
    Denmark: 12.428571428571429,
    'Dominican Republic': 7.285714285714286,
    Ecuador: 17.714285714285715,
    Egypt: 8.142857142857142,
    France: 53,
    Germany: 65.28571428571429,
    Greece: 5.142857142857143,
    Hungary: 10.857142857142858,
    India: 16.571428571428573,
    Indonesia: 11.571428571428571,
    Iran: 24.428571428571427,
    Iraq: 2.5714285714285716,
    Ireland: 21.428571428571427,
    Israel: 8.428571428571429,
    Italy: 78.71428571428571,
    Luxembourg: 4.285714285714286,
    Malaysia: 3.5714285714285716,
    Mexico: 24.714285714285715,
    Morocco: 7.571428571428571,
    Netherlands: 62.857142857142854,
    Norway: 7.857142857142857,
    Pakistan: 6.142857142857143,
    Panama: 4.142857142857143,
    Peru: 15.428571428571429,
    Philippines: 11.714285714285714,
    Poland: 15.714285714285714,
    Portugal: 24.285714285714285,
    Romania: 16.142857142857142,
    Russia: 12.142857142857142,
    'Saudi Arabia': 5.428571428571429,
    Serbia: 4.285714285714286,
    'South Korea': 3.5714285714285716,
    Spain: 125.85714285714286,
    Sweden: 38.714285714285715,
    Switzerland: 32.714285714285715,
    Turkey: 31.142857142857142,
    Ukraine: 9.714285714285714,
    'United Kingdom': 62,
    'United States of America': 72.14285714285714,
  },
  {
    date: 12,
    Algeria: 20.142857142857142,
    Argentina: 6.428571428571429,
    Austria: 16.857142857142858,
    Belgium: 92.85714285714286,
    Brazil: 38.142857142857146,
    Canada: 33.42857142857143,
    Chile: 6.857142857142857,
    China: 51.42857142857143,
    Colombia: 10.714285714285714,
    Czechia: 10,
    Denmark: 13.714285714285714,
    'Dominican Republic': 8.285714285714286,
    Ecuador: 17.428571428571427,
    Egypt: 9.428571428571429,
    France: 67.28571428571429,
    Germany: 83.28571428571429,
    Greece: 4.571428571428571,
    Hungary: 10.857142857142858,
    India: 20.428571428571427,
    Indonesia: 14.142857142857142,
    Iran: 30.571428571428573,
    Iraq: 2,
    Ireland: 23.571428571428573,
    Israel: 8.285714285714286,
    Italy: 102.85714285714286,
    Luxembourg: 3.7142857142857144,
    Malaysia: 3.7142857142857144,
    Mexico: 27.714285714285715,
    Morocco: 9,
    Netherlands: 71.85714285714286,
    Norway: 8.285714285714286,
    Pakistan: 6,
    Panama: 4.714285714285714,
    Peru: 15.714285714285714,
    Philippines: 10.857142857142858,
    Poland: 18.428571428571427,
    Portugal: 23.714285714285715,
    Romania: 17.714285714285715,
    Russia: 14.428571428571429,
    'Saudi Arabia': 5.571428571428571,
    Serbia: 4.142857142857143,
    'South Korea': 3.4285714285714284,
    Spain: 170,
    Sweden: 41.57142857142857,
    Switzerland: 41,
    Turkey: 40.142857142857146,
    'United Kingdom': 83.14285714285714,
    'United States of America': 99,
  },
  {
    date: 13,
    Algeria: 20,
    Argentina: 6.428571428571429,
    Austria: 16.857142857142858,
    Belgium: 113,
    Brazil: 45.42857142857143,
    Canada: 40.57142857142857,
    China: 56.285714285714285,
    Czechia: 10.142857142857142,
    Denmark: 15.285714285714286,
    'Dominican Republic': 8.285714285714286,
    Ecuador: 18.428571428571427,
    Egypt: 11.857142857142858,
    France: 78.14285714285714,
    Germany: 96.28571428571429,
    Greece: 4.714285714285714,
    Hungary: 11.285714285714286,
    India: 24.428571428571427,
    Indonesia: 13.142857142857142,
    Iran: 37.42857142857143,
    Iraq: 2.7142857142857144,
    Ireland: 23.857142857142858,
    Israel: 8.428571428571429,
    Italy: 124,
    Luxembourg: 3.2857142857142856,
    Malaysia: 2.857142857142857,
    Mexico: 28.857142857142858,
    Morocco: 7.428571428571429,
    Netherlands: 84.57142857142857,
    Norway: 7.142857142857143,
    Pakistan: 7,
    Panama: 4.714285714285714,
    Peru: 14.428571428571429,
    Philippines: 11.571428571428571,
    Poland: 19.714285714285715,
    Portugal: 25.142857142857142,
    Romania: 19.285714285714285,
    Russia: 16,
    Serbia: 3.142857142857143,
    'South Korea': 3.857142857142857,
    Spain: 204.57142857142858,
    Sweden: 47.285714285714285,
    Switzerland: 39.285714285714285,
    Turkey: 47.57142857142857,
    'United Kingdom': 112.28571428571429,
    'United States of America': 128.57142857142858,
  },
  {
    date: 14,
    Algeria: 18.857142857142858,
    Argentina: 6.285714285714286,
    Austria: 16,
    Belgium: 122,
    Brazil: 50,
    Canada: 46.57142857142857,
    China: 60.57142857142857,
    Czechia: 9.285714285714286,
    Denmark: 15.714285714285714,
    'Dominican Republic': 8.285714285714286,
    Ecuador: 20.142857142857142,
    Egypt: 11.428571428571429,
    France: 101.71428571428571,
    Germany: 109.14285714285714,
    Greece: 4.714285714285714,
    India: 28,
    Indonesia: 13.428571428571429,
    Iran: 46,
    Iraq: 2.142857142857143,
    Ireland: 26.142857142857142,
    Israel: 8,
    Italy: 153,
    Luxembourg: 3.2857142857142856,
    Malaysia: 2.4285714285714284,
    Mexico: 29.571428571428573,
    Morocco: 6.857142857142857,
    Netherlands: 93,
    Norway: 6.857142857142857,
    Pakistan: 8.714285714285714,
    Panama: 4.714285714285714,
    Peru: 17.571428571428573,
    Philippines: 10.714285714285714,
    Poland: 19.714285714285715,
    Portugal: 24.428571428571427,
    Romania: 17.714285714285715,
    Russia: 19.285714285714285,
    Serbia: 3.2857142857142856,
    'South Korea': 3.7142857142857144,
    Spain: 267.57142857142856,
    Sweden: 58.714285714285715,
    Switzerland: 38.714285714285715,
    Turkey: 56.142857142857146,
    'United Kingdom': 135.28571428571428,
    'United States of America': 163.71428571428572,
  },
  {
    date: 15,
    Algeria: 18.714285714285715,
    Argentina: 5.142857142857143,
    Austria: 16.428571428571427,
    Belgium: 132.85714285714286,
    Brazil: 56.285714285714285,
    Canada: 53,
    China: 66.28571428571429,
    Czechia: 10.428571428571429,
    Denmark: 16.142857142857142,
    'Dominican Republic': 13,
    Ecuador: 17.428571428571427,
    Egypt: 12.571428571428571,
    France: 132.14285714285714,
    Germany: 119,
    Greece: 4.428571428571429,
    India: 28.428571428571427,
    Indonesia: 12.714285714285714,
    Iran: 55.714285714285715,
    Iraq: 2.2857142857142856,
    Ireland: 25.142857142857142,
    Italy: 172.57142857142858,
    Luxembourg: 2.4285714285714284,
    Malaysia: 2.4285714285714284,
    Mexico: 37.857142857142854,
    Morocco: 6.571428571428571,
    Netherlands: 109,
    Pakistan: 9.857142857142858,
    Panama: 5.571428571428571,
    Peru: 19,
    Philippines: 12.714285714285714,
    Poland: 19.142857142857142,
    Portugal: 26.428571428571427,
    Romania: 20.142857142857142,
    Russia: 22.285714285714285,
    Serbia: 4.857142857142857,
    'South Korea': 3.5714285714285716,
    Spain: 315,
    Sweden: 64,
    Switzerland: 41,
    Turkey: 63.285714285714285,
    'United Kingdom': 153.28571428571428,
    'United States of America': 206.71428571428572,
  },
  {
    date: 16,
    Algeria: 16.142857142857142,
    Austria: 18.142857142857142,
    Belgium: 145.14285714285714,
    Brazil: 66.57142857142857,
    Canada: 54.57142857142857,
    China: 72.57142857142857,
    Czechia: 9.571428571428571,
    Denmark: 16.285714285714285,
    'Dominican Republic': 13,
    Ecuador: 21.714285714285715,
    Egypt: 12.285714285714286,
    France: 155.28571428571428,
    Germany: 136.14285714285714,
    Greece: 3.5714285714285716,
    India: 32.142857142857146,
    Indonesia: 12,
    Iran: 66.57142857142857,
    Iraq: 2.2857142857142856,
    Ireland: 27.285714285714285,
    Italy: 206.42857142857142,
    Malaysia: 2.2857142857142856,
    Mexico: 39.285714285714285,
    Morocco: 5.142857142857143,
    Netherlands: 116.71428571428571,
    Panama: 5.142857142857143,
    Peru: 19.428571428571427,
    Philippines: 12.285714285714286,
    Poland: 18.142857142857142,
    Portugal: 27.571428571428573,
    Romania: 22.571428571428573,
    Serbia: 5.285714285714286,
    'South Korea': 4.714285714285714,
    Spain: 405.14285714285717,
    Sweden: 73,
    Switzerland: 43.57142857142857,
    Turkey: 68.71428571428571,
    'United Kingdom': 195.28571428571428,
    'United States of America': 264.42857142857144,
  },
  {
    date: 17,
    Austria: 19.571428571428573,
    Belgium: 159.85714285714286,
    Brazil: 79.85714285714286,
    Canada: 60.285714285714285,
    China: 78.28571428571429,
    Czechia: 8.142857142857142,
    Denmark: 16.285714285714285,
    'Dominican Republic': 12.142857142857142,
    Ecuador: 21.714285714285715,
    Egypt: 13.285714285714286,
    France: 189.14285714285714,
    Germany: 139.85714285714286,
    Greece: 3.5714285714285716,
    India: 32.57142857142857,
    Indonesia: 12.428571428571429,
    Iran: 75.71428571428571,
    Iraq: 2.142857142857143,
    Ireland: 28,
    Italy: 242,
    Malaysia: 2.142857142857143,
    Mexico: 41.714285714285715,
    Morocco: 4.857142857142857,
    Netherlands: 129.28571428571428,
    Panama: 5.714285714285714,
    Philippines: 13.714285714285714,
    Poland: 20,
    Portugal: 28.571428571428573,
    Romania: 23,
    'South Korea': 4.428571428571429,
    Spain: 474.57142857142856,
    Sweden: 76.71428571428571,
    Switzerland: 43.142857142857146,
    Turkey: 73,
    'United Kingdom': 295.57142857142856,
    'United States of America': 291.14285714285717,
  },
  {
    date: 18,
    Austria: 21.571428571428573,
    Belgium: 190,
    Brazil: 91.71428571428571,
    Canada: 62.42857142857143,
    China: 84.42857142857143,
    Denmark: 15.428571428571429,
    'Dominican Republic': 11.571428571428571,
    Ecuador: 20.428571428571427,
    Egypt: 11.428571428571429,
    France: 220.71428571428572,
    Germany: 146.28571428571428,
    Greece: 2.857142857142857,
    India: 35.42857142857143,
    Indonesia: 12.142857142857142,
    Iran: 88,
    Iraq: 2,
    Ireland: 29.857142857142858,
    Italy: 267.7142857142857,
    Malaysia: 2.142857142857143,
    Morocco: 4.714285714285714,
    Netherlands: 134.42857142857142,
    Panama: 6.142857142857143,
    Philippines: 9.571428571428571,
    Portugal: 27,
    Romania: 23.142857142857142,
    'South Korea': 4.142857142857143,
    Spain: 550.8571428571429,
    Sweden: 73.42857142857143,
    Switzerland: 41.285714285714285,
    Turkey: 76.42857142857143,
    'United Kingdom': 334.7142857142857,
    'United States of America': 368.57142857142856,
  },
  {
    date: 19,
    Austria: 21.571428571428573,
    Belgium: 201.71428571428572,
    Brazil: 99.57142857142857,
    Canada: 65.28571428571429,
    China: 88.85714285714286,
    Denmark: 14.142857142857142,
    'Dominican Republic': 11.142857142857142,
    Ecuador: 21.857142857142858,
    Egypt: 11.142857142857142,
    France: 250.28571428571428,
    Germany: 161.28571428571428,
    Greece: 2.857142857142857,
    India: 34,
    Indonesia: 11.857142857142858,
    Iran: 99.57142857142857,
    Iraq: 2,
    Ireland: 31.857142857142858,
    Italy: 307.2857142857143,
    Malaysia: 2.7142857142857144,
    Netherlands: 144.57142857142858,
    Philippines: 14.714285714285714,
    Portugal: 29.142857142857142,
    Romania: 23.285714285714285,
    'South Korea': 5.428571428571429,
    Spain: 623.4285714285714,
    Sweden: 71.14285714285714,
    Switzerland: 38.285714285714285,
    Turkey: 78.85714285714286,
    'United Kingdom': 406.57142857142856,
    'United States of America': 468.2857142857143,
  },
  {
    date: 20,
    Austria: 20.857142857142858,
    Belgium: 216,
    Brazil: 98.85714285714286,
    Canada: 74.71428571428571,
    China: 114.85714285714286,
    Denmark: 13.428571428571429,
    Ecuador: 23.428571428571427,
    France: 276,
    Germany: 176.42857142857142,
    Greece: 2.7142857142857144,
    Indonesia: 15.714285714285714,
    Iran: 111.57142857142857,
    Iraq: 1.4285714285714286,
    Italy: 341.57142857142856,
    Malaysia: 2.5714285714285716,
    Netherlands: 142.14285714285714,
    Philippines: 20.714285714285715,
    Portugal: 29.857142857142858,
    Romania: 22.571428571428573,
    'South Korea': 5.571428571428571,
    Spain: 686.8571428571429,
    Sweden: 63.142857142857146,
    Switzerland: 46.714285714285715,
    Turkey: 83,
    'United Kingdom': 470.57142857142856,
    'United States of America': 584,
  },
  {
    date: 21,
    Austria: 21.142857142857142,
    Belgium: 268,
    Brazil: 105.28571428571429,
    Canada: 82.14285714285714,
    China: 106.28571428571429,
    Denmark: 14,
    Ecuador: 19.285714285714285,
    France: 309.14285714285717,
    Germany: 193.71428571428572,
    Greece: 2.7142857142857144,
    Indonesia: 17.857142857142858,
    Iran: 122.14285714285714,
    Iraq: 1.5714285714285714,
    Italy: 394.85714285714283,
    Malaysia: 2.4285714285714284,
    Netherlands: 143.28571428571428,
    Philippines: 21.714285714285715,
    Portugal: 32,
    'South Korea': 5.714285714285714,
    Spain: 736.8571428571429,
    Sweden: 63.142857142857146,
    Switzerland: 46.285714285714285,
    Turkey: 85.71428571428571,
    'United Kingdom': 529.4285714285714,
    'United States of America': 679.5714285714286,
  },
  {
    date: 22,
    Austria: 20.142857142857142,
    Belgium: 294.7142857142857,
    Brazil: 110.71428571428571,
    Canada: 97.71428571428571,
    China: 114.42857142857143,
    Denmark: 13.714285714285714,
    Ecuador: 20.857142857142858,
    France: 346.14285714285717,
    Germany: 198,
    Indonesia: 19.428571428571427,
    Iran: 131.28571428571428,
    Italy: 483.7142857142857,
    Netherlands: 151.71428571428572,
    Philippines: 22.571428571428573,
    Portugal: 31.714285714285715,
    'South Korea': 5.714285714285714,
    Spain: 784.7142857142857,
    Sweden: 73.71428571428571,
    Switzerland: 45.857142857142854,
    Turkey: 89.14285714285714,
    'United Kingdom': 566.4285714285714,
    'United States of America': 778.5714285714286,
  },
  {
    date: 23,
    Austria: 17.142857142857142,
    Belgium: 307.57142857142856,
    Brazil: 123.57142857142857,
    China: 122,
    Denmark: 13,
    Ecuador: 18.714285714285715,
    France: 385.85714285714283,
    Germany: 190.14285714285714,
    Indonesia: 25,
    Iran: 135,
    Italy: 523.5714285714286,
    Netherlands: 153.57142857142858,
    Philippines: 23.857142857142858,
    Portugal: 31.285714285714285,
    'South Korea': 5.571428571428571,
    Spain: 802.7142857142857,
    Sweden: 77.14285714285714,
    Switzerland: 41.57142857142857,
    Turkey: 92.42857142857143,
    'United Kingdom': 624.2857142857143,
    'United States of America': 901.4285714285714,
  },
  {
    date: 24,
    Austria: 16.428571428571427,
    Belgium: 324.42857142857144,
    Brazil: 133.71428571428572,
    China: 123.14285714285714,
    Denmark: 12,
    France: 401,
    Germany: 195,
    Indonesia: 27.142857142857142,
    Iran: 137.28571428571428,
    Italy: 559.8571428571429,
    Netherlands: 151,
    Philippines: 22.714285714285715,
    Portugal: 31.428571428571427,
    'South Korea': 5.857142857142857,
    Spain: 844.8571428571429,
    Switzerland: 42.714285714285715,
    Turkey: 96.85714285714286,
    'United Kingdom': 652.1428571428571,
    'United States of America': 1019.7142857142857,
  },
  {
    date: 25,
    Belgium: 303.14285714285717,
    Brazil: 140.42857142857142,
    China: 121.71428571428571,
    France: 644.5714285714286,
    Germany: 194.57142857142858,
    Indonesia: 34,
    Iran: 137,
    Italy: 616.4285714285714,
    Netherlands: 146.28571428571428,
    'South Korea': 6.857142857142857,
    Spain: 868.1428571428571,
    Switzerland: 39.142857142857146,
    Turkey: 100.85714285714286,
    'United Kingdom': 722.4285714285714,
    'United States of America': 1117,
  },
  {
    date: 26,
    Belgium: 314.2857142857143,
    China: 127.71428571428571,
    France: 749.4285714285714,
    Germany: 199,
    Indonesia: 32.714285714285715,
    Iran: 135.14285714285714,
    Italy: 646.7142857142857,
    Netherlands: 141.71428571428572,
    'South Korea': 6.428571428571429,
    Spain: 864.8571428571429,
    Switzerland: 37,
    Turkey: 105,
    'United Kingdom': 764.7142857142857,
    'United States of America': 1259.4285714285713,
  },
  {
    date: 27,
    Belgium: 333.42857142857144,
    China: 107.42857142857143,
    France: 781.7142857142857,
    Germany: 208.85714285714286,
    Indonesia: 30.857142857142858,
    Iran: 134.57142857142858,
    Italy: 679.7142857142857,
    Netherlands: 138.71428571428572,
    'South Korea': 6.142857142857143,
    Spain: 841.4285714285714,
    Switzerland: 38.285714285714285,
    'United Kingdom': 794.5714285714286,
    'United States of America': 1382.7142857142858,
  },
  {
    date: 28,
    China: 122.42857142857143,
    France: 841,
    Germany: 213.57142857142858,
    Iran: 135.71428571428572,
    Italy: 729.1428571428571,
    Netherlands: 136.57142857142858,
    'South Korea': 5.285714285714286,
    Spain: 816.4285714285714,
    Switzerland: 82.42857142857143,
    'United Kingdom': 811.1428571428571,
    'United States of America': 1519.5714285714287,
  },
  {
    date: 29,
    China: 117.57142857142857,
    France: 972.1428571428571,
    Iran: 135,
    Italy: 742.2857142857143,
    Netherlands: 120.57142857142857,
    'South Korea': 5.428571428571429,
    Spain: 801.2857142857143,
    'United Kingdom': 850.8571428571429,
    'United States of America': 1660,
  },
  {
    date: 30,
    China: 111.28571428571429,
    France: 976.7142857142857,
    Iran: 137.28571428571428,
    Italy: 757.8571428571429,
    Netherlands: 126.57142857142857,
    'South Korea': 5,
    Spain: 786,
    'United Kingdom': 849.7142857142857,
    'United States of America': 1729.5714285714287,
  },
  {
    date: 31,
    China: 117.71428571428571,
    France: 1101,
    Iran: 136.42857142857142,
    Italy: 787.7142857142857,
    Netherlands: 131.28571428571428,
    'South Korea': 4.714285714285714,
    Spain: 747.8571428571429,
    'United Kingdom': 824.4285714285714,
    'United States of America': 1780.142857142857,
  },
  {
    date: 32,
    China: 113.71428571428571,
    France: 955.7142857142857,
    Iran: 135,
    Italy: 801.4285714285714,
    'South Korea': 4.428571428571429,
    Spain: 701.1428571428571,
    'United Kingdom': 821.5714285714286,
    'United States of America': 1808.5714285714287,
  },
  {
    date: 33,
    China: 101.28571428571429,
    France: 896,
    Iran: 137.71428571428572,
    Italy: 807.4285714285714,
    'South Korea': 4,
    Spain: 658.4285714285714,
    'United States of America': 1880.2857142857142,
  },
  {
    date: 34,
    China: 89.42857142857143,
    France: 902.1428571428571,
    Iran: 137,
    Italy: 821.7142857142857,
    'South Korea': 4.142857142857143,
    Spain: 650.5714285714286,
    'United States of America': 2309.714285714286,
  },
  {
    date: 35,
    China: 78.85714285714286,
    France: 865.1428571428571,
    Iran: 132.28571428571428,
    Italy: 792.1428571428571,
    'South Korea': 5.285714285714286,
    Spain: 633.4285714285714,
    'United States of America': 2370.5714285714284,
  },
  {
    date: 36,
    China: 70,
    France: 771.5714285714286,
    Iran: 111.71428571428571,
    Italy: 762.7142857142857,
    'South Korea': 5,
    Spain: 608.2857142857143,
  },
  {
    date: 37,
    China: 61,
    France: 899.7142857142857,
    Iran: 133.57142857142858,
    Italy: 729.7142857142857,
    'South Korea': 4.857142857142857,
    Spain: 574.8571428571429,
  },
  {
    date: 38,
    China: 45.57142857142857,
    France: 815.7142857142857,
    Iran: 137.57142857142858,
    Italy: 704.8571428571429,
    'South Korea': 4.428571428571429,
    Spain: 556,
  },
  {
    date: 39,
    China: 40.142857142857146,
    Iran: 140.28571428571428,
    Italy: 671.2857142857143,
    'South Korea': 4.428571428571429,
  },
  {
    date: 40,
    China: 38,
    Iran: 139.14285714285714,
    Italy: 644.5714285714286,
    'South Korea': 4.428571428571429,
  },
  {
    date: 41,
    China: 38.285714285714285,
    Iran: 136.71428571428572,
    Italy: 623.4285714285714,
    'South Korea': 4.285714285714286,
  },
  {
    date: 42,
    China: 36.285714285714285,
    Iran: 135.71428571428572,
    Italy: 595.7142857142857,
    'South Korea': 3.5714285714285716,
  },
  {
    date: 43,
    China: 33.57142857142857,
    Iran: 153.14285714285714,
    Italy: 586.8571428571429,
    'South Korea': 3.5714285714285716,
  },
  {
    date: 44,
    China: 32.42857142857143,
    Iran: 129.28571428571428,
    Italy: 573.1428571428571,
    'South Korea': 3.142857142857143,
  },
  {
    date: 45,
    China: 29.714285714285715,
    Iran: 124.42857142857143,
    Italy: 562.8571428571429,
  },
  {
    date: 46,
    China: 27.571428571428573,
    Iran: 120.85714285714286,
    Italy: 562.8571428571429,
  },
  {
    date: 47,
    China: 25.428571428571427,
    Iran: 115.85714285714286,
    Italy: 568.2857142857143,
  },
  {
    date: 48,
    China: 22.571428571428573,
    Iran: 112,
    Italy: 555.8571428571429,
  },
  {
    date: 49,
    China: 19.285714285714285,
    Iran: 108.42857142857143,
  },
  {
    date: 50,
    China: 17.428571428571427,
  },
  {
    date: 51,
    China: 14.857142857142858,
  },
  {
    date: 52,
    China: 13.428571428571429,
  },
  {
    date: 53,
    China: 12.285714285714286,
  },
  {
    date: 54,
    China: 11.428571428571429,
  },
  {
    date: 55,
    China: 11,
  },
  {
    date: 56,
    China: 10.571428571428571,
  },
  {
    date: 57,
    China: 9.285714285714286,
  },
  {
    date: 58,
    China: 8.857142857142858,
  },
  {
    date: 59,
    China: 8.285714285714286,
  },
  {
    date: 60,
    China: 8,
  },
  {
    date: 61,
    China: 6.285714285714286,
  },
  {
    date: 62,
    China: 6,
  },
  {
    date: 63,
    China: 6.142857142857143,
  },
  {
    date: 64,
    China: 5.714285714285714,
  },
  {
    date: 65,
    China: 5.571428571428571,
  },
  {
    date: 66,
    China: 4.571428571428571,
  },
  {
    date: 67,
    China: 4,
  },
  {
    date: 68,
    China: 3.5714285714285716,
  },
  {
    date: 69,
    China: 3.5714285714285716,
  },
  {
    date: 70,
    China: 4.285714285714286,
  },
  {
    date: 71,
    China: 4.428571428571429,
  },
  {
    date: 72,
    China: 4.142857142857143,
  },
  {
    date: 73,
    China: 4.142857142857143,
  },
  {
    date: 74,
    China: 3.7142857142857144,
  },
  {
    date: 75,
    China: 3.857142857142857,
  },
  {
    date: 76,
    China: 3.2857142857142856,
  },
  {
    date: 77,
    China: 2,
  },
  {
    date: 78,
    China: 1.8571428571428572,
  },
  {
    date: 79,
    China: 1.4285714285714286,
  },
  {
    date: 80,
    China: 1.4285714285714286,
  },
  {
    date: 81,
    China: 1.4285714285714286,
  },
  {
    date: 82,
    China: 1.2857142857142858,
  },
  {
    date: 83,
    China: 1,
  },
  {
    date: 84,
    China: 0.1,
  },
];

export default dateContinuousLogSampleData;
