// @flourish/js2css/js2css.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global = global || self, global.js2css = factory());
}(this, (function () { 'use strict';

	function Stylesheet() {
		this.declarations = [];
		return this;
	}

	Stylesheet.prototype.select = function(selector) {
		if (!selector) return this;
		var declaration = new Declaration(selector, this);
		declaration.parent = this;
		this.addDeclaration(declaration);

		return declaration;
	};

	Stylesheet.prototype.addDeclaration = function(declaration) {
		this.declarations.push(declaration);
		return this;
	};

	Stylesheet.prototype.print = function() {
		var text = "";
		this.declarations.forEach(function(declaration) {
			text += declaration.selector + " {\n";
			declaration.styles.forEach(function(style) {
				text += "\t" + style[0] + ": " + style[1] + ";\n";
			});
			text += "}\n\n";
		});
		return text;
	};

	Stylesheet.prototype.clear = function() {
		this.declarations = [];
		return this;
	};

	function Declaration(selector) {
		this.selector = selector;
		this.styles = [];
		return this;
	}

	Declaration.prototype.style = function(property, _value) {
		var value = typeof value_ == "function" ? _value() : _value;
		if (value !== "" && value !== null && value !== undefined) this.styles.push([property, value]);
		return this;
	};

	Declaration.prototype.select = function(selector) {
		return this.parent.select(this.selector + " " + selector);
	};

	return Stylesheet;

})));
