// @flourish/layout/src/defaults/index.js
var DEFAULTS = Object.freeze({
	body_font: {
		name: "Source Sans Pro",
		url: "https://fonts.googleapis.com/css?family=Source+Sans+Pro:400,700"
	},
	title_font: null,
	subtitle_font: null,
	footer_font: null,

	border: {
		enabled: false,
		top: {
			width: 1,
			color: "#dddddd",
			style: "solid"
		},
		right: {
			width: 1,
			color: "#dddddd",
			style: "solid"
		},
		bottom: {
			width: 1,
			color: "#dddddd",
			style: "solid"
		},
		left: {
			width: 1,
			color: "#dddddd",
			style: "solid"
		}
	},

	layout_order: "stack-default",

	margin_top: 0.75,
	margin_right: 0.75,
	margin_bottom: 0.75,
	margin_left: 0.75,

	space_between_sections: 1,
	space_between_sections_custom: 1,

	background_color_enabled: true,
	background_color: "#ffffff",
	background_image_enabled: false,
	background_image_src: "",
	background_image_size: "cover",
	background_image_position: "center center",

	max_width: 600,
	max_width_target: "none",
	max_width_align: "center",

	breakpoint_mobile_small: 0,
	breakpoint_mobile_big: 380,
	breakpoint_tablet: 580,
	breakpoint_desktop: 1080,
	breakpoint_big_screen: 1280,

	font_color: "#333333",
	font_size_mobile_small: 62.5,
	font_size_mobile_big: 75,
	font_size_tablet: 87.5,
	font_size_desktop: 100,
	font_size_big_screen: 120,

	// header
	header_align: "left",
	header_border: "none",
	header_border_width: 1,
	header_border_color: "#dddddd",
	header_border_style: "solid",
	header_border_space: 0.5,

	header_logo_enabled: false,
	header_logo_align: "inside",
	header_logo_src: "",
	header_logo_height: 3,
	header_logo_position_inside: "left",
	header_logo_position_outside: "left",
	header_logo_margin_top: 0.25,
	header_logo_margin_right: 0.5,
	header_logo_margin_bottom: 0,
	header_logo_margin_left: 0,

	title: "",
	title_size: 1.6,
	title_size_custom: 1.6,
	title_line_height: 1.2,
	title_color: null,
	title_weight: "bold",
	title_space_above: 0,
	title_space_above_custom: 1.5,
	title_styling: false,

	subtitle: "",
	subtitle_size: 1.6,
	subtitle_size_custom: 1.6,
	subtitle_line_height: 1.2,
	subtitle_color: null,
	subtitle_weight: "normal",
	subtitle_space_above: 0,
	subtitle_space_above_custom: 1.5,
	subtitle_styling: false,

	header_text: "",
	header_text_size: 1.2,
	header_text_size_custom: 1.2,
	header_text_line_height: 1.2,
	header_text_color: null,
	header_text_weight: "normal",
	header_text_space_above: 0.5,
	header_text_space_above_custom: 1.5,

	// footer
	source_label: "Source: ",
	source_name: "",
	source_url: "",
	source_name_2: "",
	source_url_2: "",
	source_name_3: "",
	source_url_3: "",

	footer_note: "",
	footer_note_secondary: "",
	footer_text_size: 1,
	footer_text_color: null,
	footer_styling: false,

	footer_align: "justify",
	footer_align_vertical: "center",
	footer_border: "none",
	footer_border_width: 1,
	footer_border_color: "#dddddd",
	footer_border_style: "solid",
	footer_border_space: 0.5,

	footer_logo_enabled: false,
	footer_logo_src: "",
	footer_logo_src_light: "",
	footer_logo_link_url: "",
	footer_logo_height: 1.5,
	footer_logo_margin: 0.25,
	footer_logo_order: "right",

	screenreader_hide_primary: true,
	screenreader_text_primary: ""
});


function getDefaultHeight() {
	if (Flourish.fixed_height) return window.innerHeight;
	var width = window.innerWidth;
	if (width > 999) return 650;
	if (width > 599) return 575;
	return 400;
}


var DEFAULT_OVERLAY_MESSAGE = "Your web browser does not support the features used by this content. Consider updating to a modern browser.";


export { DEFAULTS, getDefaultHeight, DEFAULT_OVERLAY_MESSAGE };
