// @flourish/chart-layout/src/state/x.js
var X_DEFAULTS = Object.freeze({
	axis_position: "bottom",
	y_axis_position: "",

	numeric_scale_type: "linear",
	nice: true,
	flip: false,
	zero_axis: "auto",
	linear_min: null,
	linear_max: null,
	log_min: null,
	log_max: null,
	datetime_min: "",
	datetime_max: "",

	line_visible: true,
	line_and_tick_color: "#eeeeee",
	line_and_tick_width: 0.1,

	tick_side: "out",
	tick_mode: "auto",
	tick_number: 5,
	tick_custom: "",
	tick_length: 0.5,
	tick_padding: 0.25,
	edge_padding: 50,

	tick_label_position: "default",
	tick_label_size: 0.9,
	tick_label_color: "#aaaaaa",
	tick_label_angle: "45",
	tick_label_weight: "normal",
	tick_label_space_mode: "auto",
	tick_label_space: 10,
	tick_label_max_lines: 2,
	// Perhaps this should be nullable; then if you null it
	// it is set to 1 for continuous scales and in discrete scales
	// we do set it automatically to Math.floor(step/line_height).
	// This way we'd always favour showing more discrete ticks over
	// adding new lines.
	tick_label_line_height: 1.1,

	gridlines_visible: false,
	gridline_color: "#eeeeee",
	gridline_width: 0.1,
	gridline_style: "solid",
	gridline_category_dividers: false,
	gridline_category_dividers_extend: false,

	title_mode: "auto",
	title: "",
	title_color: "#aaaaaa",
	title_size: 1.1,
	title_weight: "normal",
	title_padding: 0.6
});


export { X_DEFAULTS };
