// @flourish/chart-layout/src/state/y.js
import { fillInDefaults } from "./common";


var Y_DEFAULTS = Object.freeze({
	axis_visible: true,

	numeric_scale_type: "linear",
	nice: true,
	flip: false,
	zero_axis: "auto",
	linear_min: null,
	linear_max: null,
	log_min: null,
	log_max: null,
	datetime_min: "",
	datetime_max: "",

	line_visible: true,
	line_and_tick_color: "#eeeeee",
	line_and_tick_width: 0.1,

	tick_side: "out",
	tick_mode: "auto",
	tick_number: 5,
	tick_custom: "",
	tick_length: 0.5,
	tick_padding: 0.25,
	edge_padding: 50,

	tick_label_position: "default",
	tick_label_size: 0.9,
	tick_label_color: "#aaaaaa",
	tick_label_angle: "0",
	tick_label_weight: "normal",
	tick_label_space_mode: "auto",
	tick_label_space: 10,
	tick_label_max_lines: 2,
	tick_label_line_height: 1.1,

	gridlines_visible: true,
	gridline_color: "#eeeeee",
	gridline_width: 0.1,
	gridline_style: "solid",
	gridline_category_dividers: false,
	gridline_category_dividers_extend: false,

	title_mode: "auto",
	title: "",
	title_color: "#aaaaaa",
	title_size: 1.1,
	title_weight: "normal",
	title_padding: 0.6
});


var Y2_DEFAULTS = Object.freeze(
	fillInDefaults({ axis_visible: false }, Y_DEFAULTS)
);


export { Y_DEFAULTS, Y2_DEFAULTS };
