// @flourish/info-popup/src/defaults.js
var DEFAULTS = {
	mode: "popup",

	// If truthy, means panel or popup are locked
	locked_id: null,

	// Popup content
	popup_is_custom: false,
	popup_custom_header: "Popup header",
	popup_custom_main: "Popup main content. Write your column names in between double curly brackets to display dynamic content.",
	popup_teaser_default: "Click to see more",

	// Popup styles
	popup_text_color: "#333333",
	popup_align: "start",
	popup_font_size: 1,
	popup_width: 12,
	popup_background: "#ffffff",
	popup_background_opacity: 1,
	popup_padding: 0.5,
	popup_radius: 0.2,
	popup_shadow: true,
	popup_pointer: true,
	popup_header_background: "#333333",
	popup_header_text_color: "#ffffff",
	popup_header_type: "block",
	popup_list_type: "table",
	popup_list_label_weight: "normal",
	popup_list_separator: ": ",

	// Panel content
	panel_is_custom: false,
	panel_custom_header: "Custom panel header",
	panel_custom_main: "Custom panel main content",
	panel_side_default_content: "<h1>Tap or click to explore</h1>",
	panel_side_fixed: false,

	// Panel layout
	panel_position: "overlay",
	panel_width_overlay: 80,
	panel_height_overlay: 80,
	panel_vertical_position: "center",
	panel_width_side: 35,
	panel_height_side: 35,
	panel_max_width: null,
	panel_max_height: null,

	// Panel styles
	panel_text_color: "#333333",
	panel_align: "start",
	panel_font_size: 1.25,
	panel_background: "#ffffff",
	panel_background_opacity: 0.9,
	panel_padding: 1.2,
	panel_radius: 0.2,
	panel_shadow: true,
	panel_open_duration: 250,
	panel_close_duration: 250,
	panel_list_type: "table",
	panel_list_label_weight: "normal",
	panel_list_separator: ": ",
};

export default DEFAULTS;
