// @financial-times/vvc/src/components/legend/props.js

import { scaleLinear, scaleOrdinal, scaleThreshold, scaleSqrt } from '@visx/scale';

export const defaultWidth = 500;
export const defaultHeight = 250;
export const defaultX = 10;
export const defaultY = 10;
export const defaultOrientation = 'vert';
export const defaultGeometry = 'rect';
export const defaultType = 'categorical';
export const defaultValueExtent = [0, 20];
export const defaultInvert = false;
export const defaultMedian = '';
export const defaultLabel = 'label not yet added';

export const defaultScale = scaleOrdinal({
    domain: [
        'Whippet',
        'Greyhound',
        'Saluki',
        'Italian Greyhound',
        'Afghan Hound',
        'Silken Windhound',
        'Kanni',
        'Galgo Español',
    ],
    range: ['#1E558C', '#94d2e6', '#1E8FCC', '#B3325D', '#FF75A3', '#D9CCC3', '#AECC70', '#F34D5B'],
});

export const defaultThresholdScale = scaleThreshold({
    domain: [1.6, 3.3, 4.3, 5.4, 7, 10.1, 14.8, 20.1],
    range: ['#d6d3ea', '#adb8e6', '#849cdb', '#5e82c8', '#3968ad', '#1b4f8d', '#0a3866'],
});

export const defaultMagnitudeScale = scaleThreshold({
    domain: [1.6, 3.3, 4.3, 5.4, 7, 10.1, 14.8, 20.1],
    range: ['#d6d3ea', '#adb8e6', '#849cdb', '#5e82c8', '#3968ad', '#1b4f8d', '#0a3866'],
});
export const defaultScaleSqrt = scaleSqrt({
    domain: [5, 20],
    range: [0, 50],
});
export const defaultRem = 16;

export default {
    defaultWidth,
    defaultHeight,
    defaultX,
    defaultY,
    defaultScale,
    defaultThresholdScale,
    defaultMedian,
    defaultScaleSqrt,
    defaultOrientation,
    defaultGeometry,
    defaultType,
    defaultInvert,
    defaultRem,
    defaultLabel,
};
