// @financial-times/vvc/src/components/legend/styles.js
/**
 * @file
 * CSS styles for VVC <Legend>
 */

import { css } from '@emotion/core';

export const baseAxisStyles = css`
  font-size: inherit !important;
`;

export const legendCategoricalStyles = css`
  ${baseAxisStyles}

  text {
    font-family: metric, MetricWeb;
    fill: #66605c;
    font-size: 16px;
  }

  line {
    stroke-width: 3px;
    stroke-linejoin: round;
    stroke-linecap: but;
  }

  svg text {
    font-feature-settings: 'tnum', 1;
    -webkit-font-feature-settings: 'tnum';
    -moz-font-feature-settings: 'tnum';
  }
`;

export const legendThresholdStyles = css`
  ${baseAxisStyles}

  text {
    font-family: metric, MetricWeb;
    fill: #66605c;
    font-size: 16px;
  }

  polygon {
    fill: #e6d9ce;
  }

  line {
    stroke: #66605c;
    stroke-width: 0.6px;
    stroke-linejoin: round;
    stroke-linecap: but;
  }

  circle {
    stroke: #66605c;
    stroke-width: 0.6px;
    stroke-linejoin: round;
    stroke-linecap: but;
  }

  svg text {
    font-feature-settings: 'tnum', 1;
    -webkit-font-feature-settings: 'tnum';
    -moz-font-feature-settings: 'tnum';
  }
`;

export const legendMagnitudeStyles = css`
  ${baseAxisStyles}

  text {
    font-family: metric, MetricWeb;
    fill: #66605c;
    font-size: 16px;
  }

  line {
    stroke: #6b6e68;
    stroke-width: 1pt;
  }

  circle {
    stroke: #6b6e68;
    stroke-width: 1pt;
    fill: none;
  }

  svg text {
    font-feature-settings: 'tnum', 1;
    -webkit-font-feature-settings: 'tnum';
    -moz-font-feature-settings: 'tnum';
  }
`;
