// @financial-times/vvc/src/components/chart-frame/styles.js
/**
 * @file
 * CSS styles for <ChartFrame />
 */

import { css } from '@emotion/core';

export const frameStyles = css`
  display: block;
  margin: 0 auto;

  text {
    font-feature-settings: 'tnum', 1;
    -moz-font-feature-settings: 'tnum';
    -webkit-font-feature-settings: 'tnum';
    -webkit-font-smoothing: antialiased;
  }
`;

export const socialChartFrameStyles = css`
  .axis text {
    fill: #8e9095;
    font-family: metric;
    font-size: 20px;
  }

  .highlighted-label {
    fill: #ffffff;
    opacity: 1;
    font-family: metric;
    font-size: 20px;
  }

  .yAxis text {
    fill: #8e9095;
    text-anchor: end;
  }

  .yAxis line {
    stroke: #47494d;
    stroke-width: 1px;
  }

  .domain {
    stroke-width: 0px;
  }

  .timelineHolder .domain {
    stroke: #ffffff;
  }

  .timelineHolder .domain {
    stroke-width: 1px;
  }

  .xAxis text {
    fill: #8e9095;
    text-anchor: middle;
  }

  .xAxis line {
    stroke: #47494d;
    stroke-width: 1px;
  }

  .baseline line {
    stroke: #8e9095;
    stroke-width: 1px;
  }

  .lines {
    stroke-linecap: square;
    stroke-linejoin: round;
    stroke-width: 4px;
    fill: none;
  }

  .annotations-holder line {
    stroke: #8e9095;
    stroke-width: 1px;
  }

  .annotations-holder path {
    stroke: #8e9095;
    stroke-width: 1px;
  }

  .annotations-holder text {
    font-size: 24px;
    fill: #ffffff;
  }

  .highlights {
    fill: #ffffff;
    opacity: 0.2;
  }

  .timeline-label,
  .legend text {
    font-family: metric;
    fill: #8e9095;
    font-size: 20px;
  }
`;

export const videoChartFrameStyles = css``;

export const printChartFrameStyles = css``;

export default frameStyles;
