// @flourish/colors/src/defaults/index.js
var DEFAULTS = Object.freeze({
	scale_type: "categorical",

	categorical_type: "palette",
	categorical_palette: ["#1D6996", "#EDAD08", "#73AF48", "#94346E", "#38A6A5", "#E17C05", "#5F4690", "#0F8554", "#6F4070", "#CC503E", "#994E95", "#666666"],
	categorical_extend: true,
	categorical_custom_palette: "",

	numeric_type: "sequential",

	sequential_palette: "Blues",
	sequential_custom_min: "#FFFFFF",
	sequential_custom_max: "#000000",
	sequential_color_space: "rgb",
	sequential_reverse: false,
	sequential_custom_domain: false,
	sequential_domain_min: 0,
	sequential_domain_max: 1,

	diverging_palette: "RdBu",
	diverging_custom_min: "#67001f",
	diverging_custom_mid: "#f7f7f7",
	diverging_custom_max: "#053061",
	diverging_color_space: "rgb",
	diverging_reverse: false,
	diverging_custom_domain: false,
	diverging_domain_min: 0,
	diverging_domain_mid: 0.5,
	diverging_domain_max: 1,

	binning: false,
	bin_count: 4,
	bin_mode: "fixed",
	bin_thresholds: ""
});


var DEFAULT_FALLBACK_VALUE = null;


export { DEFAULTS, DEFAULT_FALLBACK_VALUE };
